--- ../net/minecraft/network/packet/Packet.java
+++ ../net/minecraft/network/packet/Packet.java
@@ -16,6 +16,8 @@
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.util.IntHashMap;
 
+import org.bukkit.craftbukkit.inventory.CraftItemStack; // CraftBukkit
+
 public abstract class Packet
 {
     public static IntHashMap field_73294_l = new IntHashMap();
@@ -29,6 +31,14 @@
     public static long field_73290_p;
     public static long field_73289_q;
     public boolean field_73287_r;
+    // CraftBukkit start - Calculate packet ID once - used a bunch of times
+    private int packetID;
+
+    public Packet()
+    {
+        packetID = ((Integer) field_73291_a.get(this.getClass())).intValue();
+    }
+    // CraftBukkit end
 
     static void func_73285_a(int p_73285_0_, boolean p_73285_1_, boolean p_73285_2_, Class p_73285_3_)
     {
@@ -72,13 +82,13 @@
         }
     }
 
-    public static void func_73274_a(DataOutput p_73274_0_, byte[] p_73274_1_) throws IOException
+    public static void func_73274_a(DataOutput p_73274_0_, byte[] p_73274_1_) throws IOException   // CraftBukkit - throws IOException
     {
         p_73274_0_.writeShort(p_73274_1_.length);
         p_73274_0_.write(p_73274_1_);
     }
 
-    public static byte[] func_73280_b(DataInput p_73280_0_) throws IOException
+    public static byte[] func_73280_b(DataInput p_73280_0_) throws IOException   // CraftBukkit - throws IOException
     {
         short short1 = p_73280_0_.readShort();
 
@@ -96,10 +106,10 @@
 
     public final int func_73281_k()
     {
-        return ((Integer)field_73291_a.get(this.getClass())).intValue();
-    }
-
-    public static Packet func_73272_a(ILogAgent p_73272_0_, DataInput p_73272_1_, boolean p_73272_2_, Socket p_73272_3_) throws IOException
+        return packetID; // ((Integer) a.get(this.getClass())).intValue(); // CraftBukkit
+    }
+
+    public static Packet func_73272_a(ILogAgent p_73272_0_, DataInput p_73272_1_, boolean p_73272_2_, Socket p_73272_3_) throws IOException   // CraftBukkit - throws IOException
     {
         boolean flag1 = false;
         Packet packet = null;
@@ -135,10 +145,22 @@
         }
         catch (EOFException eofexception)
         {
-            p_73272_0_.func_98232_c("Reached end of stream for " + p_73272_3_.getInetAddress());
+            //iconsolelogmanager.severe("Reached end of stream for " + socket.getInetAddress()); // CraftBukkit - remove unnecessary logging
             return null;
         }
-
+        // CraftBukkit start
+        catch (java.net.SocketTimeoutException exception)
+        {
+            p_73272_0_.func_98233_a("Read timed out");
+            return null;
+        }
+        catch (java.net.SocketException exception)
+        {
+            p_73272_0_.func_98233_a("Connection reset");
+            return null;
+        }
+
+        // CraftBukkit end
         PacketCount.func_76118_a(j, (long)packet.func_73284_a());
         ++field_73292_n;
         field_73293_o += (long)packet.func_73284_a();
@@ -146,7 +168,7 @@
         return packet;
     }
 
-    public static void func_73266_a(Packet p_73266_0_, DataOutput p_73266_1_) throws IOException
+    public static void func_73266_a(Packet p_73266_0_, DataOutput p_73266_1_) throws IOException   // CraftBukkit - throws IOException
     {
         p_73266_1_.write(p_73266_0_.func_73281_k());
         p_73266_0_.func_73273_a(p_73266_1_);
@@ -154,7 +176,7 @@
         field_73289_q += (long)p_73266_0_.func_73284_a();
     }
 
-    public static void func_73271_a(String p_73271_0_, DataOutput p_73271_1_) throws IOException
+    public static void func_73271_a(String p_73271_0_, DataOutput p_73271_1_) throws IOException   // CraftBukkit - throws IOException
     {
         if (p_73271_0_.length() > 32767)
         {
@@ -167,7 +189,7 @@
         }
     }
 
-    public static String func_73282_a(DataInput p_73282_0_, int p_73282_1_) throws IOException
+    public static String func_73282_a(DataInput p_73282_0_, int p_73282_1_) throws IOException   // CraftBukkit - throws IOException
     {
         short short1 = p_73282_0_.readShort();
 
@@ -192,11 +214,11 @@
         }
     }
 
-    public abstract void func_73267_a(DataInput datainput) throws IOException;
-
-    public abstract void func_73273_a(DataOutput dataoutput) throws IOException;
-
-    public abstract void func_73279_a(NetHandler nethandler);
+    public abstract void func_73267_a(DataInput datainput) throws IOException; // CraftBukkit - throws IOException
+
+    public abstract void func_73273_a(DataOutput dataoutput) throws IOException; // CraftBukkit - throws IOException
+
+    public abstract void func_73279_a(NetHandler connection);
 
     public abstract int func_73284_a();
 
@@ -212,7 +234,7 @@
 
     public boolean func_73277_a_()
     {
-        return false;
+        return this instanceof Packet3Chat && !((Packet3Chat) this).field_73476_b.startsWith("/"); // CraftBukkit - async chat
     }
 
     public String toString()
@@ -221,7 +243,7 @@
         return s;
     }
 
-    public static ItemStack func_73276_c(DataInput p_73276_0_) throws IOException
+    public static ItemStack func_73276_c(DataInput p_73276_0_) throws IOException   // CraftBukkit - throws IOException
     {
         ItemStack itemstack = null;
         short short1 = p_73276_0_.readShort();
@@ -232,14 +254,22 @@
             short short2 = p_73276_0_.readShort();
             itemstack = new ItemStack(short1, b0, short2);
             itemstack.field_77990_d = func_73283_d(p_73276_0_);
+
+            // CraftBukkit start
+            if (itemstack.field_77990_d != null)
+            {
+                CraftItemStack.setItemMeta(itemstack, CraftItemStack.getItemMeta(itemstack));
+            }
+
+            // CraftBukkit end
         }
 
         return itemstack;
     }
 
-    public static void func_73270_a(ItemStack p_73270_0_, DataOutput p_73270_1_) throws IOException
-    {
-        if (p_73270_0_ == null)
+    public static void func_73270_a(ItemStack p_73270_0_, DataOutput p_73270_1_) throws IOException   // CraftBukkit - throws IOException
+    {
+        if (p_73270_0_ == null || p_73270_0_.func_77973_b() == null)   // CraftBukkit - NPE fix itemstack.getItem()
         {
             p_73270_1_.writeShort(-1);
         }
@@ -259,7 +289,7 @@
         }
     }
 
-    public static NBTTagCompound func_73283_d(DataInput p_73283_0_) throws IOException
+    public static NBTTagCompound func_73283_d(DataInput p_73283_0_) throws IOException   // CraftBukkit - throws IOException
     {
         short short1 = p_73283_0_.readShort();
 
@@ -275,7 +305,7 @@
         }
     }
 
-    protected static void func_73275_a(NBTTagCompound p_73275_0_, DataOutput p_73275_1_) throws IOException
+    protected static void func_73275_a(NBTTagCompound p_73275_0_, DataOutput p_73275_1_) throws IOException   // CraftBukkit - throws IOException
     {
         if (p_73275_0_ == null)
         {
