--- ../net/minecraft/server/ServerListenThread.java
+++ ../net/minecraft/server/ServerListenThread.java
@@ -21,7 +21,9 @@
     private final InetAddress field_71772_g;
     private final int field_71778_h;
 
-    public ServerListenThread(NetworkListenThread p_i3384_1_, InetAddress p_i3384_2_, int p_i3384_3_) throws IOException
+    long connectionThrottle; // CraftBukkit
+
+    public ServerListenThread(NetworkListenThread p_i3384_1_, InetAddress p_i3384_2_, int p_i3384_3_) throws IOException   // CraftBukkit - added throws
     {
         super("Listen thread");
         this.field_71771_f = p_i3384_1_;
@@ -48,7 +50,7 @@
                 catch (Exception exception)
                 {
                     netloginhandler.func_72527_a("Internal server error");
-                    this.field_71771_f.func_71746_d().func_98033_al().func_98235_b("Failed to handle packet for " + netloginhandler.func_72528_e() + ": " + exception, exception);
+                    this.field_71771_f.func_71746_d().func_98033_al().func_98235_b("Failed to handle packet for " + netloginhandler.func_72528_e() + ": " + exception, (Throwable) exception);
                 }
 
                 if (netloginhandler.field_72539_c)
@@ -68,12 +70,37 @@
             try
             {
                 Socket socket = this.field_71774_e.accept();
+                // CraftBukkit start - connection throttle
+                InetAddress address = socket.getInetAddress();
+                long currentTime = System.currentTimeMillis();
+
+                if (((MinecraftServer) this.field_71771_f.func_71746_d()).server == null)
+                {
+                    socket.close();
+                    continue;
+                }
+
+                connectionThrottle = ((MinecraftServer) this.field_71771_f.func_71746_d()).server.getConnectionThrottle();
+
+                synchronized (this.field_71776_c)
+                {
+                    if (this.field_71776_c.containsKey(address) && !"127.0.0.1".equals(address.getHostAddress()) && currentTime - ((Long) this.field_71776_c.get(address)).longValue() < connectionThrottle)
+                    {
+                        this.field_71776_c.put(address, Long.valueOf(currentTime));
+                        socket.close();
+                        continue;
+                    }
+
+                    this.field_71776_c.put(address, Long.valueOf(currentTime));
+                }
+
+                // CraftBukkit end
                 NetLoginHandler netloginhandler = new NetLoginHandler(this.field_71771_f.func_71746_d(), socket, "Connection #" + this.field_71773_d++);
                 this.func_71764_a(netloginhandler);
             }
             catch (IOException ioexception)
             {
-                ioexception.printStackTrace();
+                this.field_71771_f.func_71746_d().func_98033_al().func_98236_b("DSCT: " + ioexception.getMessage()); // CraftBukkit
             }
         }
 
@@ -84,7 +111,7 @@
     {
         if (p_71764_1_ == null)
         {
-            throw new IllegalArgumentException("Got null pendingconnection!");
+            throw new IllegalArgumentException("Got null netloginhandler!");
         }
         else
         {
