--- ../net/minecraft/world/chunk/storage/ExtendedBlockStorage.java
+++ ../net/minecraft/world/chunk/storage/ExtendedBlockStorage.java
@@ -2,7 +2,6 @@
 
 import net.minecraft.block.Block;
 import net.minecraft.world.chunk.NibbleArray;
-
 public class ExtendedBlockStorage
 {
     private int field_76684_a;
@@ -27,6 +26,29 @@
         }
     }
 
+    // CraftBukkit start
+    public ExtendedBlockStorage(int y, boolean flag, byte[] blkIds, byte[] extBlkIds)
+    {
+        this.field_76684_a = y;
+        this.field_76680_d = blkIds;
+
+        if (extBlkIds != null)
+        {
+            this.field_76681_e = new NibbleArray(extBlkIds, 4);
+        }
+
+        this.field_76678_f = new NibbleArray(this.field_76680_d.length, 4);
+        this.field_76679_g = new NibbleArray(this.field_76680_d.length, 4);
+
+        if (flag)
+        {
+            this.field_76685_h = new NibbleArray(this.field_76680_d.length, 4);
+        }
+
+        this.func_76672_e();
+    }
+    // CraftBukkit end
+
     public int func_76656_a(int p_76656_1_, int p_76656_2_, int p_76656_3_)
     {
         int l = this.field_76680_d[p_76656_2_ << 8 | p_76656_3_ << 4 | p_76656_1_] & 255;
@@ -133,6 +155,95 @@
 
     public void func_76672_e()
     {
+        // CraftBukkit start - Optimize for speed
+        byte[] blkIds = this.field_76680_d;
+        int cntNonEmpty = 0;
+        int cntTicking = 0;
+
+        if (this.field_76681_e == null)   // No extended block IDs?  Don't waste time messing with them
+        {
+            for (int off = 0; off < blkIds.length; off++)
+            {
+                int l = blkIds[off] & 0xFF;
+
+                if (l > 0)
+                {
+                    if (Block.field_71973_m[l] == null)
+                    {
+                        blkIds[off] = 0;
+                    }
+                    else
+                    {
+                        ++cntNonEmpty;
+
+                        if (Block.field_71973_m[l].func_71881_r())
+                        {
+                            ++cntTicking;
+                        }
+                    }
+                }
+            }
+        }
+        else
+        {
+            byte[] ext = this.field_76681_e.field_76585_a;
+
+            for (int off = 0, off2 = 0; off < blkIds.length;)
+            {
+                byte extid = ext[off2];
+                int l = (blkIds[off] & 0xFF) | ((extid & 0xF) << 8); // Even data
+
+                if (l > 0)
+                {
+                    if (Block.field_71973_m[l] == null)
+                    {
+                        blkIds[off] = 0;
+                        ext[off2] &= 0xF0;
+                    }
+                    else
+                    {
+                        ++cntNonEmpty;
+
+                        if (Block.field_71973_m[l].func_71881_r())
+                        {
+                            ++cntTicking;
+                        }
+                    }
+                }
+
+                off++;
+                l = (blkIds[off] & 0xFF) | ((extid & 0xF0) << 4); // Odd data
+
+                if (l > 0)
+                {
+                    if (Block.field_71973_m[l] == null)
+                    {
+                        blkIds[off] = 0;
+                        ext[off2] &= 0x0F;
+                    }
+                    else
+                    {
+                        ++cntNonEmpty;
+
+                        if (Block.field_71973_m[l].func_71881_r())
+                        {
+                            ++cntTicking;
+                        }
+                    }
+                }
+
+                off++;
+                off2++;
+            }
+        }
+
+        this.field_76682_b = cntNonEmpty;
+        this.field_76683_c = cntTicking;
+    }
+
+    public void old_recalcBlockCounts()
+    {
+        // CraftBukkit end
         this.field_76682_b = 0;
         this.field_76683_c = 0;
 
@@ -197,26 +308,70 @@
 
     public void func_76664_a(byte[] p_76664_1_)
     {
-        this.field_76680_d = p_76664_1_;
+        this.field_76680_d = this.validateByteArray(p_76664_1_); // CraftBukkit - Validate data
     }
 
     public void func_76673_a(NibbleArray p_76673_1_)
     {
-        this.field_76681_e = p_76673_1_;
+        // CraftBukkit start - Don't hang on to an empty nibble array
+        boolean empty = true;
+
+        for (int i = 0; i < p_76673_1_.field_76585_a.length; i++)
+        {
+            if (p_76673_1_.field_76585_a[i] != 0)
+            {
+                empty = false;
+                break;
+            }
+        }
+
+        if (empty)
+        {
+            return;
+        }
+
+        // CraftBukkit end
+        this.field_76681_e = this.validateNibbleArray(p_76673_1_); // CraftBukkit - Validate data
     }
 
     public void func_76668_b(NibbleArray p_76668_1_)
     {
-        this.field_76678_f = p_76668_1_;
+        this.field_76678_f = this.validateNibbleArray(p_76668_1_); // CraftBukkit - Validate data
     }
 
     public void func_76659_c(NibbleArray p_76659_1_)
     {
-        this.field_76679_g = p_76659_1_;
+        this.field_76679_g = this.validateNibbleArray(p_76659_1_); // CraftBukkit - Validate data
     }
 
     public void func_76666_d(NibbleArray p_76666_1_)
     {
-        this.field_76685_h = p_76666_1_;
-    }
+        this.field_76685_h = this.validateNibbleArray(p_76666_1_); // CraftBukkit - Validate data
+    }
+
+    // CraftBukkit start - Validate array lengths
+    private NibbleArray validateNibbleArray(NibbleArray nibbleArray)
+    {
+        if (nibbleArray != null && nibbleArray.field_76585_a.length < 2048)
+        {
+            byte[] newArray = new byte[2048];
+            System.arraycopy(nibbleArray.field_76585_a, 0, newArray, 0, ((nibbleArray.field_76585_a.length > 2048) ? 2048 : nibbleArray.field_76585_a.length));
+            nibbleArray = new NibbleArray(newArray, 4);
+        }
+
+        return nibbleArray;
+    }
+
+    private byte[] validateByteArray(byte[] byteArray)
+    {
+        if (byteArray != null && byteArray.length < 4096)
+        {
+            byte[] newArray = new byte[4096];
+            System.arraycopy(byteArray, 0, newArray, 0, byteArray.length);
+            byteArray = newArray;
+        }
+
+        return byteArray;
+    }
+    // CraftBukkit end
 }
