--- ../net/minecraft/entity/player/EntityPlayerMP.java
+++ ../net/minecraft/entity/player/EntityPlayerMP.java
@@ -4,7 +4,6 @@
 import java.io.DataOutputStream;
 import java.io.IOException;
 import java.util.ArrayList;
-import java.util.Arrays;
 import java.util.Collection;
 import java.util.Iterator;
 import java.util.LinkedList;
@@ -65,7 +64,6 @@
 import net.minecraft.network.packet.Packet8UpdateHealth;
 import net.minecraft.potion.PotionEffect;
 import net.minecraft.scoreboard.Score;
-import net.minecraft.scoreboard.ScoreObjective;
 import net.minecraft.scoreboard.ScoreObjectiveCriteria;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.stats.AchievementList;
@@ -81,8 +79,10 @@
 import net.minecraft.tileentity.TileEntitySign;
 import net.minecraft.util.ChatMessageComponent;
 import net.minecraft.util.ChunkCoordinates;
+import net.minecraft.util.CombatTracker;
 import net.minecraft.util.DamageSource;
 import net.minecraft.util.EntityDamageSource;
+import net.minecraft.util.FoodStats;
 import net.minecraft.util.MathHelper;
 import net.minecraft.util.ReportedException;
 import net.minecraft.village.MerchantRecipeList;
@@ -91,6 +91,17 @@
 import net.minecraft.world.World;
 import net.minecraft.world.WorldServer;
 import net.minecraft.world.chunk.Chunk;
+
+// CraftBukkit start
+import org.bukkit.Bukkit;
+import org.bukkit.WeatherType;
+import org.bukkit.craftbukkit.CraftWorld;
+import org.bukkit.craftbukkit.entity.CraftPlayer;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.event.inventory.InventoryType;
+import org.bukkit.event.player.PlayerTeleportEvent.TeleportCause;
+// CraftBukkit end
 
 public class EntityPlayerMP extends EntityPlayer implements ICrafting
 {
@@ -106,15 +117,25 @@
     private float field_71149_ch = -1.0E8F;
     private int field_71146_ci = -99999999;
     private boolean field_71147_cj = true;
-    private int field_71144_ck = -99999999;
-    private int field_71145_cl = 60;
+    public int field_71144_ck = -99999999; // CraftBukkit - private -> public
+    public int field_71145_cl = 60; // CraftBukkit - private -> public
     private int field_71142_cm;
     private int field_71143_cn;
     private boolean field_71140_co = true;
-    public int field_71139_cq;
+    private int field_71139_cq;
     public boolean field_71137_h;
     public int field_71138_i;
     public boolean field_71136_j;
+    // CraftBukkit start
+    public String displayName;
+    public String listName;
+    public org.bukkit.Location compassTarget;
+    public int newExp = 0;
+    public int newLevel = 0;
+    public int newTotalExp = 0;
+    public boolean keepLevel = false;
+    public double maxHealthCache;
+    // CraftBukkit end
 
     public EntityPlayerMP(MinecraftServer p_i1523_1_, World p_i1523_2_, String p_i1523_3_, ItemInWorldManager p_i1523_4_)
     {
@@ -144,6 +165,13 @@
         {
             this.func_70107_b(this.field_70165_t, this.field_70163_u + 1.0D, this.field_70161_v);
         }
+
+        // CraftBukkit start
+        this.displayName = this.field_71092_bJ;
+        this.listName = this.field_71092_bJ;
+        // this.canPickUpLoot = true; TODO
+        this.maxHealthCache = this.func_110138_aP();
+        // CraftBukkit end
     }
 
     public void func_70037_a(NBTTagCompound p_70037_1_)
@@ -161,13 +189,52 @@
                 this.field_71134_c.func_73076_a(EnumGameType.func_77146_a(p_70037_1_.func_74762_e("playerGameType")));
             }
         }
+
+        this.getBukkitEntity().readExtraData(p_70037_1_); // CraftBukkit
     }
 
     public void func_70014_b(NBTTagCompound p_70014_1_)
     {
         super.func_70014_b(p_70014_1_);
         p_70014_1_.func_74768_a("playerGameType", this.field_71134_c.func_73081_b().func_77148_a());
-    }
+        this.getBukkitEntity().setExtraData(p_70014_1_); // CraftBukkit
+    }
+
+    // CraftBukkit start - World fallback code, either respawn location or global spawn
+    public void func_70029_a(World world)
+    {
+        super.func_70029_a(world);
+
+        if (world == null)
+        {
+            this.field_70128_L = false;
+            ChunkCoordinates position = null;
+
+            if (this.spawnWorld != null && !this.spawnWorld.equals(""))
+            {
+                CraftWorld cworld = (CraftWorld) Bukkit.getServer().getWorld(this.spawnWorld);
+
+                if (cworld != null && this.func_70997_bJ() != null)
+                {
+                    world = cworld.getHandle();
+                    position = EntityPlayer.func_71056_a(cworld.getHandle(), this.func_70997_bJ(), false);
+                }
+            }
+
+            if (world == null || position == null)
+            {
+                world = ((CraftWorld) Bukkit.getServer().getWorlds().get(0)).getHandle();
+                position = world.func_72861_E();
+            }
+
+            this.field_70170_p = world;
+            this.func_70107_b(position.field_71574_a + 0.5, position.field_71572_b, position.field_71573_c + 0.5);
+        }
+
+        this.field_71093_bK = ((WorldServer) this.field_70170_p).dimension;
+        this.field_71134_c.func_73080_a((WorldServer) world);
+    }
+    // CraftBukkit end
 
     public void func_82242_a(int p_82242_1_)
     {
@@ -196,11 +263,20 @@
         --this.field_71145_cl;
         this.field_71070_bA.func_75142_b();
 
-        if (!this.field_70170_p.field_72995_K && !this.field_71070_bA.func_75145_c(this))
+        if (!this.field_70170_p.field_72995_K && !this.field_71070_bA.func_75145_c((EntityPlayer) this))
         {
             this.func_71053_j();
             this.field_71070_bA = this.field_71069_bz;
         }
+
+        // CraftBukkit start - Check inventory status every tick
+        if (!this.field_71070_bA.func_75145_c(this))   // Should be stillValid
+        {
+            this.func_71053_j();
+            this.field_71070_bA = this.field_71069_bz;
+        }
+
+        // CraftBukkit end
 
         while (!this.field_71130_g.isEmpty())
         {
@@ -231,8 +307,11 @@
 
                 if (chunkcoordintpair != null && this.field_70170_p.func_72899_e(chunkcoordintpair.field_77276_a << 4, 0, chunkcoordintpair.field_77275_b << 4))
                 {
-                    arraylist.add(this.field_70170_p.func_72964_e(chunkcoordintpair.field_77276_a, chunkcoordintpair.field_77275_b));
-                    arraylist1.addAll(((WorldServer)this.field_70170_p).func_73049_a(chunkcoordintpair.field_77276_a * 16, 0, chunkcoordintpair.field_77275_b * 16, chunkcoordintpair.field_77276_a * 16 + 16, 256, chunkcoordintpair.field_77275_b * 16 + 16));
+                    // CraftBukkit start - Get tile entities directly from the chunk instead of the world
+                    Chunk chunk = this.field_70170_p.func_72964_e(chunkcoordintpair.field_77276_a, chunkcoordintpair.field_77275_b);
+                    arraylist.add(chunk);
+                    arraylist1.addAll(chunk.field_76648_i.values());
+                    // CraftBukkit end
                 }
             }
 
@@ -281,7 +360,8 @@
 
             if (this.func_110143_aJ() != this.field_71149_ch || this.field_71146_ci != this.field_71100_bB.func_75116_a() || this.field_71100_bB.func_75115_e() == 0.0F != this.field_71147_cj)
             {
-                this.field_71135_a.func_72567_b(new Packet8UpdateHealth(this.func_110143_aJ(), this.field_71100_bB.func_75116_a(), this.field_71100_bB.func_75115_e()));
+                // CraftBukkit - Optionally scale health
+                this.field_71135_a.func_72567_b(new Packet8UpdateHealth(this.getBukkitEntity().getScaledHealth(), this.field_71100_bB.func_75116_a(), this.field_71100_bB.func_75115_e()));
                 this.field_71149_ch = this.func_110143_aJ();
                 this.field_71146_ci = this.field_71100_bB.func_75116_a();
                 this.field_71147_cj = this.field_71100_bB.func_75115_e() == 0.0F;
@@ -290,21 +370,37 @@
             if (this.func_110143_aJ() + this.func_110139_bj() != this.field_130068_bO)
             {
                 this.field_130068_bO = this.func_110143_aJ() + this.func_110139_bj();
-                Collection collection = this.func_96123_co().func_96520_a(ScoreObjectiveCriteria.field_96638_f);
-                Iterator iterator = collection.iterator();
-
-                while (iterator.hasNext())
-                {
-                    ScoreObjective scoreobjective = (ScoreObjective)iterator.next();
-                    this.func_96123_co().func_96529_a(this.func_70023_ak(), scoreobjective).func_96651_a(Arrays.asList(new EntityPlayer[] {this}));
-                }
-            }
+                // CraftBukkit - Update ALL the scores!
+                this.field_70170_p.getServer().getScoreboardManager().updateAllScoresForList(ScoreObjectiveCriteria.field_96638_f, this.func_70023_ak(), com.google.common.collect.ImmutableList.of(this));
+            }
+
+            // CraftBukkit start - Force max health updates
+            if (this.maxHealthCache != this.func_110138_aP())
+            {
+                this.getBukkitEntity().updateScaledHealth();
+            }
+
+            // CraftBukkit end
 
             if (this.field_71067_cb != this.field_71144_ck)
             {
                 this.field_71144_ck = this.field_71067_cb;
                 this.field_71135_a.func_72567_b(new Packet43Experience(this.field_71106_cc, this.field_71067_cb, this.field_71068_ca));
             }
+
+            // CraftBukkit start
+            if (this.oldLevel == -1)
+            {
+                this.oldLevel = this.field_71068_ca;
+            }
+
+            if (this.oldLevel != this.field_71068_ca)
+            {
+                CraftEventFactory.callPlayerLevelChangeEvent(this.field_70170_p.getServer().getPlayer((EntityPlayerMP) this), this.oldLevel, this.field_71068_ca);
+                this.oldLevel = this.field_71068_ca;
+            }
+
+            // CraftBukkit end
         }
         catch (Throwable throwable)
         {
@@ -317,20 +413,74 @@
 
     public void func_70645_a(DamageSource p_70645_1_)
     {
-        this.field_71133_b.func_71203_ab().func_92062_k(this.func_110142_aN().func_94546_b());
-
-        if (!this.field_70170_p.func_82736_K().func_82766_b("keepInventory"))
-        {
-            this.field_71071_by.func_70436_m();
-        }
-
-        Collection collection = this.field_70170_p.func_96441_U().func_96520_a(ScoreObjectiveCriteria.field_96642_c);
+        // CraftBukkit start
+        if (this.field_70128_L)
+        {
+            return;
+        }
+
+        java.util.List<org.bukkit.inventory.ItemStack> loot = new java.util.ArrayList<org.bukkit.inventory.ItemStack>();
+        boolean keepInventory = this.field_70170_p.func_82736_K().func_82766_b("keepInventory");
+
+        if (!keepInventory)
+        {
+            for (int i = 0; i < this.field_71071_by.field_70462_a.length; ++i)
+            {
+                if (this.field_71071_by.field_70462_a[i] != null)
+                {
+                    loot.add(CraftItemStack.asCraftMirror(this.field_71071_by.field_70462_a[i]));
+                }
+            }
+
+            for (int i = 0; i < this.field_71071_by.field_70460_b.length; ++i)
+            {
+                if (this.field_71071_by.field_70460_b[i] != null)
+                {
+                    loot.add(CraftItemStack.asCraftMirror(this.field_71071_by.field_70460_b[i]));
+                }
+            }
+        }
+
+        ChatMessageComponent chatmessagecomponent = this.func_110142_aN().func_94546_b();
+        String deathmessage = chatmessagecomponent.toString();
+        org.bukkit.event.entity.PlayerDeathEvent event = CraftEventFactory.callPlayerDeathEvent(this, loot, deathmessage);
+        String deathMessage = event.getDeathMessage();
+
+        if (deathMessage != null && deathMessage.length() > 0)
+        {
+            if (deathMessage.equals(chatmessagecomponent.toString()))
+            {
+                this.field_71133_b.func_71203_ab().func_92062_k(chatmessagecomponent);
+            }
+            else
+            {
+                this.field_71133_b.func_71203_ab().func_92062_k(ChatMessageComponent.func_111066_d(event.getDeathMessage()));
+            }
+        }
+
+        // CraftBukkit - we clean the player's inventory after the EntityDeathEvent is called so plugins can get the exact state of the inventory.
+        if (!keepInventory)
+        {
+            for (int i = 0; i < this.field_71071_by.field_70462_a.length; ++i)
+            {
+                this.field_71071_by.field_70462_a[i] = null;
+            }
+
+            for (int i = 0; i < this.field_71071_by.field_70460_b.length; ++i)
+            {
+                this.field_71071_by.field_70460_b[i] = null;
+            }
+        }
+
+        this.func_71053_j();
+        // CraftBukkit end
+        // CraftBukkit - Get our scores instead
+        Collection<Score> collection = this.field_70170_p.getServer().getScoreboardManager().getScoreboardScores(ScoreObjectiveCriteria.field_96642_c, this.func_70023_ak(), new java.util.ArrayList<Score>());
         Iterator iterator = collection.iterator();
 
         while (iterator.hasNext())
         {
-            ScoreObjective scoreobjective = (ScoreObjective)iterator.next();
-            Score score = this.func_96123_co().func_96529_a(this.func_70023_ak(), scoreobjective);
+            Score score = (Score) iterator.next(); // CraftBukkit - Use our scores instead
             score.func_96648_a();
         }
 
@@ -352,7 +502,8 @@
         }
         else
         {
-            boolean flag = this.field_71133_b.func_71262_S() && this.field_71133_b.func_71219_W() && "fall".equals(p_70097_1_.field_76373_n);
+            // CraftBukkit - this.server.getPvP() -> this.world.pvpMode
+            boolean flag = this.field_71133_b.func_71262_S() && this.field_70170_p.pvpMode && "fall".equals(p_70097_1_.field_76373_n);
 
             if (!flag && this.field_71145_cl > 0 && p_70097_1_ != DamageSource.field_76380_i)
             {
@@ -387,14 +538,15 @@
 
     public boolean func_96122_a(EntityPlayer p_96122_1_)
     {
-        return !this.field_71133_b.func_71219_W() ? false : super.func_96122_a(p_96122_1_);
+        // CraftBukkit - this.server.getPvP() -> this.world.pvpMode
+        return !this.field_70170_p.pvpMode ? false : super.func_96122_a(p_96122_1_);
     }
 
     public void func_71027_c(int p_71027_1_)
     {
         if (this.field_71093_bK == 1 && p_71027_1_ == 1)
         {
-            this.func_71029_a(AchievementList.field_76003_C);
+            this.func_71029_a((StatBase) AchievementList.field_76003_C);
             this.field_70170_p.func_72900_e(this);
             this.field_71136_j = true;
             this.field_71135_a.func_72567_b(new Packet70GameEvent(4, 0));
@@ -403,22 +555,28 @@
         {
             if (this.field_71093_bK == 0 && p_71027_1_ == 1)
             {
-                this.func_71029_a(AchievementList.field_76002_B);
-                ChunkCoordinates chunkcoordinates = this.field_71133_b.func_71218_a(p_71027_1_).func_73054_j();
-
-                if (chunkcoordinates != null)
-                {
-                    this.field_71135_a.func_72569_a((double)chunkcoordinates.field_71574_a, (double)chunkcoordinates.field_71572_b, (double)chunkcoordinates.field_71573_c, 0.0F, 0.0F);
+                this.func_71029_a((StatBase) AchievementList.field_76002_B);
+                // CraftBukkit start - Rely on custom portal management
+                /*
+                ChunkCoordinates chunkcoordinates = this.server.getWorldServer(i).getDimensionSpawn();
+
+                if (chunkcoordinates != null) {
+                    this.playerConnection.a((double) chunkcoordinates.x, (double) chunkcoordinates.y, (double) chunkcoordinates.z, 0.0F, 0.0F);
                 }
 
-                p_71027_1_ = 1;
+                i = 1;
+                */
+                // CraftBukkit end
             }
             else
             {
-                this.func_71029_a(AchievementList.field_76029_x);
-            }
-
-            this.field_71133_b.func_71203_ab().func_72356_a(this, p_71027_1_);
+                this.func_71029_a((StatBase) AchievementList.field_76029_x);
+            }
+
+            // CraftBukkit start
+            TeleportCause cause = (this.field_71093_bK == 1 || p_71027_1_ == 1) ? TeleportCause.END_PORTAL : TeleportCause.NETHER_PORTAL;
+            this.field_71133_b.func_71203_ab().changeDimension(this, p_71027_1_, cause);
+            // CraftBukkit end
             this.field_71144_ck = -1;
             this.field_71149_ch = -1.0F;
             this.field_71146_ci = -1;
@@ -451,7 +609,7 @@
         if (enumstatus == EnumStatus.OK)
         {
             Packet17Sleep packet17sleep = new Packet17Sleep(this, 0, p_71018_1_, p_71018_2_, p_71018_3_);
-            this.func_71121_q().func_73039_n().func_72784_a(this, packet17sleep);
+            this.func_71121_q().func_73039_n().func_72784_a((Entity) this, (Packet) packet17sleep);
             this.field_71135_a.func_72569_a(this.field_70165_t, this.field_70163_u, this.field_70161_v, this.field_70177_z, this.field_70125_A);
             this.field_71135_a.func_72567_b(packet17sleep);
         }
@@ -461,6 +619,11 @@
 
     public void func_70999_a(boolean p_70999_1_, boolean p_70999_2_, boolean p_70999_3_)
     {
+        if (this.fauxSleeping && !this.field_71083_bS)
+        {
+            return;    // CraftBukkit - Can't leave bed if not in one!
+        }
+
         if (this.func_70608_bn())
         {
             this.func_71121_q().func_73039_n().func_72789_b(this, new Packet18Animation(this, 3));
@@ -476,9 +639,25 @@
 
     public void func_70078_a(Entity p_70078_1_)
     {
-        super.func_70078_a(p_70078_1_);
-        this.field_71135_a.func_72567_b(new Packet39AttachEntity(0, this, this.field_70154_o));
-        this.field_71135_a.func_72569_a(this.field_70165_t, this.field_70163_u, this.field_70161_v, this.field_70177_z, this.field_70125_A);
+        // CraftBukkit start
+        this.setPassengerOf(p_70078_1_);
+    }
+
+    public void setPassengerOf(Entity entity)
+    {
+        // mount(null) doesn't really fly for overloaded methods,
+        // so this method is needed
+        Entity currentVehicle = this.field_70154_o;
+        super.setPassengerOf(entity);
+
+        // Check if the vehicle actually changed.
+        if (currentVehicle != this.field_70154_o)
+        {
+            this.field_71135_a.func_72567_b(new Packet39AttachEntity(0, this, this.field_70154_o));
+            this.field_71135_a.func_72569_a(this.field_70165_t, this.field_70163_u, this.field_70161_v, this.field_70177_z, this.field_70125_A);
+        }
+
+        // CraftBukkit end
     }
 
     protected void func_70064_a(double p_70064_1_, boolean p_70064_3_) {}
@@ -492,39 +671,67 @@
     {
         if (p_71014_1_ instanceof TileEntitySign)
         {
-            ((TileEntitySign)p_71014_1_).func_142010_a(this);
+            ((TileEntitySign) p_71014_1_).func_142010_a((EntityPlayer) this);
             this.field_71135_a.func_72567_b(new Packet133TileEditorOpen(0, p_71014_1_.field_70329_l, p_71014_1_.field_70330_m, p_71014_1_.field_70327_n));
         }
     }
 
-    public void func_71117_bO()
+    public int nextContainerCounter()   // CraftBukkit - private void -> public int
     {
         this.field_71139_cq = this.field_71139_cq % 100 + 1;
+        return this.field_71139_cq; // CraftBukkit
     }
 
     public void func_71058_b(int p_71058_1_, int p_71058_2_, int p_71058_3_)
     {
-        this.func_71117_bO();
+        // CraftBukkit start - Inventory open hook
+        Container container = CraftEventFactory.callInventoryOpenEvent(this, new ContainerWorkbench(this.field_71071_by, this.field_70170_p, p_71058_1_, p_71058_2_, p_71058_3_));
+
+        if (container == null)
+        {
+            return;
+        }
+
+        // CraftBukkit end
+        this.nextContainerCounter();
         this.field_71135_a.func_72567_b(new Packet100OpenWindow(this.field_71139_cq, 1, "Crafting", 9, true));
-        this.field_71070_bA = new ContainerWorkbench(this.field_71071_by, this.field_70170_p, p_71058_1_, p_71058_2_, p_71058_3_);
+        this.field_71070_bA = container; // CraftBukkit - Use container we passed to event
         this.field_71070_bA.field_75152_c = this.field_71139_cq;
         this.field_71070_bA.func_75132_a(this);
     }
 
     public void func_71002_c(int p_71002_1_, int p_71002_2_, int p_71002_3_, String p_71002_4_)
     {
-        this.func_71117_bO();
+        // CraftBukkit start - Inventory open hook
+        Container container = CraftEventFactory.callInventoryOpenEvent(this, new ContainerEnchantment(this.field_71071_by, this.field_70170_p, p_71002_1_, p_71002_2_, p_71002_3_));
+
+        if (container == null)
+        {
+            return;
+        }
+
+        // CraftBukkit end
+        this.nextContainerCounter();
         this.field_71135_a.func_72567_b(new Packet100OpenWindow(this.field_71139_cq, 4, p_71002_4_ == null ? "" : p_71002_4_, 9, p_71002_4_ != null));
-        this.field_71070_bA = new ContainerEnchantment(this.field_71071_by, this.field_70170_p, p_71002_1_, p_71002_2_, p_71002_3_);
+        this.field_71070_bA = container; // CraftBukkit - Use container we passed to event
         this.field_71070_bA.field_75152_c = this.field_71139_cq;
         this.field_71070_bA.func_75132_a(this);
     }
 
     public void func_82244_d(int p_82244_1_, int p_82244_2_, int p_82244_3_)
     {
-        this.func_71117_bO();
+        // CraftBukkit start - Inventory open hook
+        Container container = CraftEventFactory.callInventoryOpenEvent(this, new ContainerRepair(this.field_71071_by, this.field_70170_p, p_82244_1_, p_82244_2_, p_82244_3_, this));
+
+        if (container == null)
+        {
+            return;
+        }
+
+        // CraftBukkit end
+        this.nextContainerCounter();
         this.field_71135_a.func_72567_b(new Packet100OpenWindow(this.field_71139_cq, 8, "Repairing", 9, true));
-        this.field_71070_bA = new ContainerRepair(this.field_71071_by, this.field_70170_p, p_82244_1_, p_82244_2_, p_82244_3_, this);
+        this.field_71070_bA = container; // CraftBukkit - Use container we passed to event
         this.field_71070_bA.field_75152_c = this.field_71139_cq;
         this.field_71070_bA.func_75132_a(this);
     }
@@ -536,71 +743,143 @@
             this.func_71053_j();
         }
 
-        this.func_71117_bO();
+        // CraftBukkit start - Inventory open hook
+        Container container = CraftEventFactory.callInventoryOpenEvent(this, new ContainerChest(this.field_71071_by, p_71007_1_));
+
+        if (container == null)
+        {
+            return;
+        }
+
+        // CraftBukkit end
+        this.nextContainerCounter();
         this.field_71135_a.func_72567_b(new Packet100OpenWindow(this.field_71139_cq, 0, p_71007_1_.func_70303_b(), p_71007_1_.func_70302_i_(), p_71007_1_.func_94042_c()));
-        this.field_71070_bA = new ContainerChest(this.field_71071_by, p_71007_1_);
+        this.field_71070_bA = container; // CraftBukkit - Use container we passed to event
         this.field_71070_bA.field_75152_c = this.field_71139_cq;
         this.field_71070_bA.func_75132_a(this);
     }
 
     public void func_94064_a(TileEntityHopper p_94064_1_)
     {
-        this.func_71117_bO();
+        // CraftBukkit start - Inventory open hook
+        Container container = CraftEventFactory.callInventoryOpenEvent(this, new ContainerHopper(this.field_71071_by, p_94064_1_));
+
+        if (container == null)
+        {
+            return;
+        }
+
+        // CraftBukkit end
+        this.nextContainerCounter();
         this.field_71135_a.func_72567_b(new Packet100OpenWindow(this.field_71139_cq, 9, p_94064_1_.func_70303_b(), p_94064_1_.func_70302_i_(), p_94064_1_.func_94042_c()));
-        this.field_71070_bA = new ContainerHopper(this.field_71071_by, p_94064_1_);
+        this.field_71070_bA = container; // CraftBukkit - Use container we passed to event
         this.field_71070_bA.field_75152_c = this.field_71139_cq;
         this.field_71070_bA.func_75132_a(this);
     }
 
     public void func_96125_a(EntityMinecartHopper p_96125_1_)
     {
-        this.func_71117_bO();
+        // CraftBukkit start - Inventory open hook
+        Container container = CraftEventFactory.callInventoryOpenEvent(this, new ContainerHopper(this.field_71071_by, p_96125_1_));
+
+        if (container == null)
+        {
+            return;
+        }
+
+        // CraftBukkit end
+        this.nextContainerCounter();
         this.field_71135_a.func_72567_b(new Packet100OpenWindow(this.field_71139_cq, 9, p_96125_1_.func_70303_b(), p_96125_1_.func_70302_i_(), p_96125_1_.func_94042_c()));
-        this.field_71070_bA = new ContainerHopper(this.field_71071_by, p_96125_1_);
+        this.field_71070_bA = container; // CraftBukkit - Use container we passed to event
         this.field_71070_bA.field_75152_c = this.field_71139_cq;
         this.field_71070_bA.func_75132_a(this);
     }
 
     public void func_71042_a(TileEntityFurnace p_71042_1_)
     {
-        this.func_71117_bO();
+        // CraftBukkit start - Inventory open hook
+        Container container = CraftEventFactory.callInventoryOpenEvent(this, new ContainerFurnace(this.field_71071_by, p_71042_1_));
+
+        if (container == null)
+        {
+            return;
+        }
+
+        // CraftBukkit end
+        this.nextContainerCounter();
         this.field_71135_a.func_72567_b(new Packet100OpenWindow(this.field_71139_cq, 2, p_71042_1_.func_70303_b(), p_71042_1_.func_70302_i_(), p_71042_1_.func_94042_c()));
-        this.field_71070_bA = new ContainerFurnace(this.field_71071_by, p_71042_1_);
+        this.field_71070_bA = container; // CraftBukkit - Use container we passed to event
         this.field_71070_bA.field_75152_c = this.field_71139_cq;
         this.field_71070_bA.func_75132_a(this);
     }
 
     public void func_71006_a(TileEntityDispenser p_71006_1_)
     {
-        this.func_71117_bO();
+        // CraftBukkit start - Inventory open hook
+        Container container = CraftEventFactory.callInventoryOpenEvent(this, new ContainerDispenser(this.field_71071_by, p_71006_1_));
+
+        if (container == null)
+        {
+            return;
+        }
+
+        // CraftBukkit end
+        this.nextContainerCounter();
         this.field_71135_a.func_72567_b(new Packet100OpenWindow(this.field_71139_cq, p_71006_1_ instanceof TileEntityDropper ? 10 : 3, p_71006_1_.func_70303_b(), p_71006_1_.func_70302_i_(), p_71006_1_.func_94042_c()));
-        this.field_71070_bA = new ContainerDispenser(this.field_71071_by, p_71006_1_);
+        this.field_71070_bA = container; // CraftBukkit - Use container we passed to event
         this.field_71070_bA.field_75152_c = this.field_71139_cq;
         this.field_71070_bA.func_75132_a(this);
     }
 
     public void func_71017_a(TileEntityBrewingStand p_71017_1_)
     {
-        this.func_71117_bO();
+        // CraftBukkit start - Inventory open hook
+        Container container = CraftEventFactory.callInventoryOpenEvent(this, new ContainerBrewingStand(this.field_71071_by, p_71017_1_));
+
+        if (container == null)
+        {
+            return;
+        }
+
+        // CraftBukkit end
+        this.nextContainerCounter();
         this.field_71135_a.func_72567_b(new Packet100OpenWindow(this.field_71139_cq, 5, p_71017_1_.func_70303_b(), p_71017_1_.func_70302_i_(), p_71017_1_.func_94042_c()));
-        this.field_71070_bA = new ContainerBrewingStand(this.field_71071_by, p_71017_1_);
+        this.field_71070_bA = container; // CraftBukkit - Use container we passed to event
         this.field_71070_bA.field_75152_c = this.field_71139_cq;
         this.field_71070_bA.func_75132_a(this);
     }
 
     public void func_82240_a(TileEntityBeacon p_82240_1_)
     {
-        this.func_71117_bO();
+        // CraftBukkit start - Inventory open hook
+        Container container = CraftEventFactory.callInventoryOpenEvent(this, new ContainerBeacon(this.field_71071_by, p_82240_1_));
+
+        if (container == null)
+        {
+            return;
+        }
+
+        // CraftBukkit end
+        this.nextContainerCounter();
         this.field_71135_a.func_72567_b(new Packet100OpenWindow(this.field_71139_cq, 7, p_82240_1_.func_70303_b(), p_82240_1_.func_70302_i_(), p_82240_1_.func_94042_c()));
-        this.field_71070_bA = new ContainerBeacon(this.field_71071_by, p_82240_1_);
+        this.field_71070_bA = container; // CraftBukkit - Use container we passed to event
         this.field_71070_bA.field_75152_c = this.field_71139_cq;
         this.field_71070_bA.func_75132_a(this);
     }
 
     public void func_71030_a(IMerchant p_71030_1_, String p_71030_2_)
     {
-        this.func_71117_bO();
-        this.field_71070_bA = new ContainerMerchant(this.field_71071_by, p_71030_1_, this.field_70170_p);
+        // CraftBukkit start - Inventory open hook
+        Container container = CraftEventFactory.callInventoryOpenEvent(this, new ContainerMerchant(this.field_71071_by, p_71030_1_, this.field_70170_p));
+
+        if (container == null)
+        {
+            return;
+        }
+
+        // CraftBukkit end
+        this.nextContainerCounter();
+        this.field_71070_bA = container; // CraftBukkit - Use container we passed to event
         this.field_71070_bA.field_75152_c = this.field_71139_cq;
         this.field_71070_bA.func_75132_a(this);
         InventoryMerchant inventorymerchant = ((ContainerMerchant)this.field_71070_bA).func_75174_d();
@@ -626,14 +905,24 @@
 
     public void func_110298_a(EntityHorse p_110298_1_, IInventory p_110298_2_)
     {
+        // CraftBukkit start - Inventory open hook
+        Container container = CraftEventFactory.callInventoryOpenEvent(this, new ContainerHorseInventory(this.field_71071_by, p_110298_2_, p_110298_1_));
+
+        if (container == null)
+        {
+            return;
+        }
+
+        // CraftBukkit end
+
         if (this.field_71070_bA != this.field_71069_bz)
         {
             this.func_71053_j();
         }
 
-        this.func_71117_bO();
+        this.nextContainerCounter();
         this.field_71135_a.func_72567_b(new Packet100OpenWindow(this.field_71139_cq, 11, p_110298_2_.func_70303_b(), p_110298_2_.func_70302_i_(), p_110298_2_.func_94042_c(), p_110298_1_.field_70157_k));
-        this.field_71070_bA = new ContainerHorseInventory(this.field_71071_by, p_110298_2_, p_110298_1_);
+        this.field_71070_bA = container; // CraftBukkit - Use container we passed to event
         this.field_71070_bA.field_75152_c = this.field_71139_cq;
         this.field_71070_bA.func_75132_a(this);
     }
@@ -658,6 +947,14 @@
     {
         this.field_71135_a.func_72567_b(new Packet104WindowItems(p_71110_1_.field_75152_c, p_71110_2_));
         this.field_71135_a.func_72567_b(new Packet103SetSlot(-1, -1, this.field_71071_by.func_70445_o()));
+
+        // CraftBukkit start - Send a Set Slot to update the crafting result slot
+        if (java.util.EnumSet.of(InventoryType.CRAFTING, InventoryType.WORKBENCH).contains(p_71110_1_.getBukkitView().getType()))
+        {
+            this.field_71135_a.func_72567_b(new Packet103SetSlot(p_71110_1_.field_75152_c, 0, p_71110_1_.func_75139_a(0).func_75211_c()));
+        }
+
+        // CraftBukkit end
     }
 
     public void func_71112_a(Container p_71112_1_, int p_71112_2_, int p_71112_3_)
@@ -667,6 +964,7 @@
 
     public void func_71053_j()
     {
+        CraftEventFactory.handleInventoryCloseEvent(this); // CraftBukkit
         this.field_71135_a.func_72567_b(new Packet101CloseWindow(this.field_71070_bA.field_75152_c));
         this.func_71128_l();
     }
@@ -681,7 +979,7 @@
 
     public void func_71128_l()
     {
-        this.field_71070_bA.func_75134_a(this);
+        this.field_71070_bA.func_75134_a((EntityPlayer) this);
         this.field_71070_bA = this.field_71069_bz;
     }
 
@@ -731,6 +1029,7 @@
     public void func_71118_n()
     {
         this.field_71149_ch = -1.0E8F;
+        this.field_71144_ck = -1; // CraftBukkit - Added to reset
     }
 
     public void func_71035_c(String p_71035_1_)
@@ -861,7 +1160,7 @@
 
     public void func_71115_a(String p_71115_1_, int p_71115_2_)
     {
-        String s1 = p_71115_1_ + "\u0000" + p_71115_2_;
+        String s1 = p_71115_1_ + "\0" + p_71115_2_; // CraftBukkit - fix decompile error
         this.field_71135_a.func_72567_b(new Packet250CustomPayload("MC|TPack", s1.getBytes()));
     }
 
@@ -869,4 +1168,103 @@
     {
         return new ChunkCoordinates(MathHelper.func_76128_c(this.field_70165_t), MathHelper.func_76128_c(this.field_70163_u + 0.5D), MathHelper.func_76128_c(this.field_70161_v));
     }
+
+    // CraftBukkit start
+    public long timeOffset = 0;
+    public boolean relativeTime = true;
+
+    public long getPlayerTime()
+    {
+        if (this.relativeTime)
+        {
+            // Adds timeOffset to the current server time.
+            return this.field_70170_p.func_72820_D() + this.timeOffset;
+        }
+        else
+        {
+            // Adds timeOffset to the beginning of this day.
+            return this.field_70170_p.func_72820_D() - (this.field_70170_p.func_72820_D() % 24000) + this.timeOffset;
+        }
+    }
+
+    public WeatherType weather = null;
+
+    public WeatherType getPlayerWeather()
+    {
+        return this.weather;
+    }
+
+    public void setPlayerWeather(WeatherType type, boolean plugin)
+    {
+        if (!plugin && this.weather != null)
+        {
+            return;
+        }
+
+        if (plugin)
+        {
+            this.weather = type;
+        }
+
+        this.field_71135_a.func_72567_b(new Packet70GameEvent(type == WeatherType.DOWNFALL ? 1 : 2, 0));
+    }
+
+    public void resetPlayerWeather()
+    {
+        this.weather = null;
+        this.setPlayerWeather(this.field_70170_p.func_72912_H().func_76059_o() ? WeatherType.DOWNFALL : WeatherType.CLEAR, false);
+    }
+
+    @Override
+    public String toString()
+    {
+        return super.toString() + "(" + this.field_71092_bJ + " at " + this.field_70165_t + "," + this.field_70163_u + "," + this.field_70161_v + ")";
+    }
+
+    public void reset()
+    {
+        float exp = 0;
+        boolean keepInventory = this.field_70170_p.func_82736_K().func_82766_b("keepInventory");
+
+        if (this.keepLevel || keepInventory)
+        {
+            exp = this.field_71106_cc;
+            this.newTotalExp = this.field_71067_cb;
+            this.newLevel = this.field_71068_ca;
+        }
+
+        this.func_70606_j(this.func_110138_aP());
+        this.field_70151_c = 0;
+        this.field_70143_R = 0;
+        this.field_71100_bB = new FoodStats();
+        this.field_71068_ca = this.newLevel;
+        this.field_71067_cb = this.newTotalExp;
+        this.field_71106_cc = 0;
+        this.field_70725_aQ = 0;
+        this.func_70674_bp(); // Should be removeAllEffects.
+        this.field_70752_e = true;
+        this.field_71070_bA = this.field_71069_bz;
+        this.field_70717_bb = null;
+        this.field_70755_b = null;
+        this.field_94063_bt = new CombatTracker(this);
+        this.field_71144_ck = -1;
+
+        if (this.keepLevel || keepInventory)
+        {
+            this.field_71106_cc = exp;
+        }
+        else
+        {
+            this.func_71023_q(this.newExp);
+        }
+
+        this.keepLevel = false;
+    }
+
+    @Override
+    public CraftPlayer getBukkitEntity()
+    {
+        return (CraftPlayer) super.getBukkitEntity();
+    }
+    // CraftBukkit end
 }
