--- ../net/minecraft/entity/passive/EntityWolf.java
+++ ../net/minecraft/entity/passive/EntityWolf.java
@@ -30,7 +30,6 @@
 import net.minecraft.util.DamageSource;
 import net.minecraft.util.MathHelper;
 import net.minecraft.world.World;
-
 public class EntityWolf extends EntityTameable
 {
     private float field_70926_e;
@@ -134,7 +133,8 @@
 
     protected String func_70639_aQ()
     {
-        return this.func_70919_bu() ? "mob.wolf.growl" : (this.field_70146_Z.nextInt(3) == 0 ? (this.func_70909_n() && this.field_70180_af.func_111145_d(18) < 10.0F ? "mob.wolf.whine" : "mob.wolf.panting") : "mob.wolf.bark");
+        // CraftBukkit - (getInt(18) < 10) -> (getInt(18) < this.getMaxHealth() / 2)
+        return this.func_70919_bu() ? "mob.wolf.growl" : (this.field_70146_Z.nextInt(3) == 0 ? (this.func_70909_n() && this.field_70180_af.func_111145_d(18) < (this.func_110138_aP() / 2) ? "mob.wolf.whine" : "mob.wolf.panting") : "mob.wolf.bark");
     }
 
     protected String func_70621_aR()
@@ -349,13 +349,14 @@
 
             if (!this.field_70170_p.field_72995_K)
             {
-                if (this.field_70146_Z.nextInt(3) == 0)
+                // CraftBukkit - added event call and isCancelled check.
+                if (this.field_70146_Z.nextInt(3) == 0 && !org.bukkit.craftbukkit.event.CraftEventFactory.callEntityTameEvent(this, p_70085_1_).isCancelled())
                 {
                     this.func_70903_f(true);
                     this.func_70778_a((PathEntity)null);
                     this.func_70624_b((EntityLivingBase)null);
                     this.field_70911_d.func_75270_a(true);
-                    this.func_70606_j(20.0F);
+                    this.func_70606_j(this.func_110138_aP()); // CraftBukkit - 20.0 -> getMaxHealth()
                     this.func_70910_a(p_70085_1_.func_70005_c_());
                     this.func_70908_e(true);
                     this.field_70170_p.func_72960_a(this, (byte)7);
@@ -466,7 +467,7 @@
 
     protected boolean func_70692_ba()
     {
-        return !this.func_70909_n() && this.field_70173_aa > 2400;
+        return !this.func_70909_n(); // CraftBukkit
     }
 
     public boolean func_142018_a(EntityLivingBase p_142018_1_, EntityLivingBase p_142018_2_)
@@ -477,7 +478,7 @@
             {
                 EntityWolf entitywolf = (EntityWolf)p_142018_1_;
 
-                if (entitywolf.func_70909_n() && entitywolf.func_130012_q() == p_142018_2_)
+                if (entitywolf.func_70909_n() && entitywolf.func_70902_q() == p_142018_2_)
                 {
                     return false;
                 }
@@ -495,4 +496,7 @@
     {
         return this.func_70879_a(p_90011_1_);
     }
-}
+    public net.minecraft.entity.Entity func_70902_q(){
+        return func_130012_q();
+    }
+}