--- ../net/minecraft/entity/EntityLiving.java
+++ ../net/minecraft/entity/EntityLiving.java
@@ -24,11 +24,15 @@
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.nbt.NBTTagFloat;
 import net.minecraft.nbt.NBTTagList;
+import net.minecraft.network.packet.Packet;
 import net.minecraft.network.packet.Packet39AttachEntity;
 import net.minecraft.pathfinding.PathNavigate;
 import net.minecraft.util.MathHelper;
 import net.minecraft.world.World;
 import net.minecraft.world.WorldServer;
+
+//CraftBukkit start
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public abstract class EntityLiving extends EntityLivingBase
 {
@@ -39,14 +43,14 @@
     private EntityJumpHelper field_70767_i;
     private EntityBodyHelper field_70762_j;
     private PathNavigate field_70699_by;
-    public final EntityAITasks field_70714_bg;
+    protected final EntityAITasks field_70714_bg;
     protected final EntityAITasks field_70715_bh;
     private EntityLivingBase field_70696_bz;
     private EntitySenses field_70723_bA;
     private ItemStack[] field_82182_bS = new ItemStack[5];
-    protected float[] field_82174_bp = new float[5];
-    private boolean field_82172_bs;
-    private boolean field_82179_bU;
+    public float[] field_82174_bp = new float[5]; // CraftBukkit - protected -> public
+    public boolean field_82172_bs; // CraftBukkit - private -> public
+    public boolean field_82179_bU = !func_70692_ba(); // CraftBukkit - private -> public
     protected float field_70698_bv;
     private Entity field_70776_bF;
     protected int field_70700_bx;
@@ -226,6 +230,8 @@
 
     protected void func_70628_a(boolean p_70628_1_, int p_70628_2_)
     {
+        // CraftBukkit start - Whole method
+        List<org.bukkit.inventory.ItemStack> loot = new java.util.ArrayList<org.bukkit.inventory.ItemStack>();
         int j = this.func_70633_aT();
 
         if (j > 0)
@@ -237,11 +243,30 @@
                 k += this.field_70146_Z.nextInt(p_70628_2_ + 1);
             }
 
-            for (int l = 0; l < k; ++l)
-            {
-                this.func_70025_b(j, 1);
-            }
-        }
+            if (k > 0)
+            {
+                loot.add(new org.bukkit.inventory.ItemStack(j, k));
+            }
+        }
+
+        // Determine rare item drops and add them to the loot
+        if (this.field_70718_bc > 0)
+        {
+            int k = this.field_70146_Z.nextInt(200) - p_70628_2_;
+
+            if (k < 5)
+            {
+                ItemStack itemstack = this.func_70600_l(k <= 0 ? 1 : 0);
+
+                if (itemstack != null)
+                {
+                    loot.add(org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(itemstack));
+                }
+            }
+        }
+
+        CraftEventFactory.callEntityDeathEvent(this, loot); // raise event even for those times when the entity does not drop loot
+        // CraftBukkit end
     }
 
     public void func_70014_b(NBTTagCompound p_70014_1_)
@@ -303,6 +328,22 @@
         super.func_70037_a(p_70037_1_);
         this.func_98053_h(p_70037_1_.func_74767_n("CanPickUpLoot"));
         this.field_82179_bU = p_70037_1_.func_74767_n("PersistenceRequired");
+        // CraftBukkit start - If looting or persistence is false only use it if it was set after we started using it
+        boolean data = p_70037_1_.func_74767_n("CanPickUpLoot");
+
+        if (isLevelAtLeast(p_70037_1_, 1) || data)
+        {
+            this.field_82172_bs = data;
+        }
+
+        data = p_70037_1_.func_74767_n("PersistenceRequired");
+
+        if (isLevelAtLeast(p_70037_1_, 1) || data)
+        {
+            this.field_82179_bU = data;
+        }
+
+        // CraftBukkit end
 
         if (p_70037_1_.func_74764_b("CustomName") && p_70037_1_.func_74779_i("CustomName").length() > 0)
         {
@@ -475,12 +516,12 @@
                 double d2 = entityplayer.field_70161_v - this.field_70161_v;
                 double d3 = d0 * d0 + d1 * d1 + d2 * d2;
 
-                if (this.func_70692_ba() && d3 > 16384.0D)
+                if (d3 > 16384.0D)   // CraftBukkit - remove isTypeNotPersistent() check
                 {
                     this.func_70106_y();
                 }
 
-                if (this.field_70708_bq > 600 && this.field_70146_Z.nextInt(800) == 0 && d3 > 1024.0D && this.func_70692_ba())
+                if (this.field_70708_bq > 600 && this.field_70146_Z.nextInt(800) == 0 && d3 > 1024.0D)   // CraftBukkit - remove isTypeNotPersistent() check
                 {
                     this.func_70106_y();
                 }
@@ -551,7 +592,7 @@
         {
             this.func_70625_a(this.field_70776_bF, 10.0F, (float)this.func_70646_bf());
 
-            if (this.field_70700_bx-- <= 0 || this.field_70776_bF.field_70128_L || this.field_70776_bF.func_70068_e(this) > (double)(f * f))
+            if (this.field_70700_bx-- <= 0 || this.field_70776_bF.field_70128_L || this.field_70776_bF.func_70068_e((Entity) this) > (double)(f * f))
             {
                 this.field_70776_bF = null;
             }
@@ -1025,7 +1066,7 @@
 
             if (!this.field_70170_p.field_72995_K && p_110160_1_ && this.field_70170_p instanceof WorldServer)
             {
-                ((WorldServer)this.field_70170_p).func_73039_n().func_72784_a(this, new Packet39AttachEntity(1, this, (Entity)null));
+                ((WorldServer) this.field_70170_p).func_73039_n().func_72784_a((Entity) this, (Packet)(new Packet39AttachEntity(1, this, (Entity) null)));
             }
         }
     }
@@ -1052,7 +1093,7 @@
 
         if (!this.field_70170_p.field_72995_K && p_110162_2_ && this.field_70170_p instanceof WorldServer)
         {
-            ((WorldServer)this.field_70170_p).func_73039_n().func_72784_a(this, new Packet39AttachEntity(1, this, this.field_110168_bw));
+            ((WorldServer) this.field_70170_p).func_73039_n().func_72784_a((Entity) this, (Packet)(new Packet39AttachEntity(1, this, this.field_110168_bw)));
         }
     }
 
@@ -1099,4 +1140,7 @@
 
         this.field_110170_bx = null;
     }
-}
+    public boolean func_70692_ba_CodeFix_Public(){
+        return func_70692_ba();
+    }
+}