--- ../net/minecraft/inventory/ContainerRepair.java
+++ ../net/minecraft/inventory/ContainerRepair.java
@@ -11,7 +11,10 @@
 import net.minecraft.item.ItemEnchantedBook;
 import net.minecraft.item.ItemStack;
 import net.minecraft.world.World;
+
 import org.apache.commons.lang3.StringUtils;
+
+import org.bukkit.craftbukkit.inventory.CraftInventoryView; // CraftBukkit
 
 public class ContainerRepair extends Container
 {
@@ -25,9 +28,14 @@
     private int field_82856_l;
     private String field_82857_m;
     private final EntityPlayer field_82855_n;
+    // CraftBukkit start
+    private CraftInventoryView bukkitEntity = null;
+    private InventoryPlayer player;
+    // CraftBukkit end
 
     public ContainerRepair(InventoryPlayer p_i1800_1_, World p_i1800_2_, int p_i1800_3_, int p_i1800_4_, int p_i1800_5_, EntityPlayer p_i1800_6_)
     {
+        this.player = p_i1800_1_; // CraftBukkit
         this.field_82860_h = p_i1800_2_;
         this.field_82861_i = p_i1800_3_;
         this.field_82858_j = p_i1800_4_;
@@ -35,7 +43,7 @@
         this.field_82855_n = p_i1800_6_;
         this.func_75146_a(new Slot(this.field_82853_g, 0, 27, 47));
         this.func_75146_a(new Slot(this.field_82853_g, 1, 76, 47));
-        this.func_75146_a(new ContainerRepairINNER2(this, this.field_82852_f, 2, 134, 47, p_i1800_2_, p_i1800_3_, p_i1800_4_, p_i1800_5_));
+        this.func_75146_a((Slot)(new ContainerRepairINNER2(this, this.field_82852_f, 2, 134, 47, p_i1800_2_, p_i1800_3_, p_i1800_4_, p_i1800_5_)));
         int l;
 
         for (l = 0; l < 3; ++l)
@@ -368,6 +376,11 @@
 
     public boolean func_75145_c(EntityPlayer p_75145_1_)
     {
+        if (!this.checkReachable)
+        {
+            return true;    // CraftBukkit
+        }
+
         return this.field_82860_h.func_72798_a(this.field_82861_i, this.field_82858_j, this.field_82859_k) != Block.field_82510_ck.field_71990_ca ? false : p_75145_1_.func_70092_e((double)this.field_82861_i + 0.5D, (double)this.field_82858_j + 0.5D, (double)this.field_82859_k + 0.5D) <= 64.0D;
     }
 
@@ -452,4 +465,18 @@
     {
         return p_82849_0_.field_82856_l;
     }
+
+    // CraftBukkit start
+    public CraftInventoryView getBukkitView()
+    {
+        if (bukkitEntity != null)
+        {
+            return bukkitEntity;
+        }
+
+        org.bukkit.craftbukkit.inventory.CraftInventory inventory = new org.bukkit.craftbukkit.inventory.CraftInventoryAnvil(this.field_82853_g, this.field_82852_f);
+        bukkitEntity = new CraftInventoryView(this.player.field_70458_d.getBukkitEntity(), inventory, this);
+        return bukkitEntity;
+    }
+    // CraftBukkit end
 }
