--- ../net/minecraft/entity/ai/EntityAITarget.java
+++ ../net/minecraft/entity/ai/EntityAITarget.java
@@ -1,5 +1,6 @@
 package net.minecraft.entity.ai;
 
+import org.apache.commons.lang3.StringUtils;
 import net.minecraft.entity.EntityCreature;
 import net.minecraft.entity.EntityLivingBase;
 import net.minecraft.entity.EntityOwnable;
@@ -9,7 +10,11 @@
 import net.minecraft.pathfinding.PathEntity;
 import net.minecraft.pathfinding.PathPoint;
 import net.minecraft.util.MathHelper;
-import org.apache.commons.lang3.StringUtils;
+
+// CraftBukkit start
+import org.bukkit.craftbukkit.entity.CraftEntity;
+import org.bukkit.event.entity.EntityTargetEvent;
+// CraftBukkit end
 
 public abstract class EntityAITarget extends EntityAIBase
 {
@@ -154,6 +159,51 @@
                     }
                 }
 
+                // CraftBukkit start - Check all the different target goals for the reason, default to RANDOM_TARGET
+                EntityTargetEvent.TargetReason reason = EntityTargetEvent.TargetReason.RANDOM_TARGET;
+
+                if (this instanceof EntityAIDefendVillage)
+                {
+                    reason = EntityTargetEvent.TargetReason.DEFEND_VILLAGE;
+                }
+                else if (this instanceof EntityAIHurtByTarget)
+                {
+                    reason = EntityTargetEvent.TargetReason.TARGET_ATTACKED_ENTITY;
+                }
+                else if (this instanceof EntityAINearestAttackableTarget)
+                {
+                    if (p_75296_1_ instanceof EntityPlayer)
+                    {
+                        reason = EntityTargetEvent.TargetReason.CLOSEST_PLAYER;
+                    }
+                }
+                else if (this instanceof EntityAIOwnerHurtByTarget)
+                {
+                    reason = EntityTargetEvent.TargetReason.TARGET_ATTACKED_OWNER;
+                }
+                else if (this instanceof EntityAIOwnerHurtTarget)
+                {
+                    reason = EntityTargetEvent.TargetReason.OWNER_ATTACKED_TARGET;
+                }
+
+                org.bukkit.event.entity.EntityTargetLivingEntityEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callEntityTargetLivingEvent(this.field_75299_d, p_75296_1_, reason);
+
+                if (event.isCancelled() || event.getTarget() == null)
+                {
+                    this.field_75299_d.func_70624_b(null);
+                    return false;
+                }
+                else if (p_75296_1_.getBukkitEntity() != event.getTarget())
+                {
+                    this.field_75299_d.func_70624_b((EntityLivingBase)((CraftEntity) event.getTarget()).getHandle());
+                }
+
+                if (this.field_75299_d instanceof EntityCreature)
+                {
+                    ((EntityCreature) this.field_75299_d).field_70789_a = ((CraftEntity) event.getTarget()).getHandle();
+                }
+
+                // CraftBukkit end
                 return true;
             }
         }
