--- ../net/minecraft/entity/item/EntityPainting.java
+++ ../net/minecraft/entity/item/EntityPainting.java
@@ -17,6 +17,7 @@
     public EntityPainting(World p_i1599_1_)
     {
         super(p_i1599_1_);
+        this.field_70522_e = EnumArt.values()[this.field_70146_Z.nextInt(EnumArt.values().length)]; // CraftBukkit - generate a non-null painting
     }
 
     public EntityPainting(World p_i1600_1_, int p_i1600_2_, int p_i1600_3_, int p_i1600_4_, int p_i1600_5_)
