--- ../net/minecraft/entity/monster/EntitySlime.java
+++ ../net/minecraft/entity/monster/EntitySlime.java
@@ -1,5 +1,11 @@
 package net.minecraft.entity.monster;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.entity.CraftEntity;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityTargetEvent;
+import org.bukkit.event.entity.SlimeSplitEvent;
+import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityLiving;
 import net.minecraft.entity.SharedMonsterAttributes;
 import net.minecraft.entity.player.EntityPlayer;
@@ -11,6 +17,7 @@
 import net.minecraft.world.WorldType;
 import net.minecraft.world.biome.BiomeGenBase;
 import net.minecraft.world.chunk.Chunk;
+// CraftBukkit end
 
 public class EntitySlime extends EntityLiving implements IMob
 {
@@ -18,6 +25,7 @@
     public float field_70811_b;
     public float field_70812_c;
     private int field_70810_d;
+    private Entity lastTarget; // CraftBukkit
 
     public EntitySlime(World p_i1742_1_)
     {
@@ -34,7 +42,8 @@
         this.field_70180_af.func_75682_a(16, new Byte((byte)1));
     }
 
-    protected void func_70799_a(int p_70799_1_)
+    // CraftBukkit - protected -> public
+    public void func_70799_a(int p_70799_1_)
     {
         this.field_70180_af.func_75692_b(16, new Byte((byte)p_70799_1_));
         this.func_70105_a(0.6F * (float)p_70799_1_, 0.6F * (float)p_70799_1_);
@@ -121,18 +130,37 @@
     protected void func_70626_be()
     {
         this.func_70623_bb();
-        EntityPlayer entityplayer = this.field_70170_p.func_72856_b(this, 16.0D);
-
-        if (entityplayer != null)
-        {
-            this.func_70625_a(entityplayer, 10.0F, 20.0F);
+        // CraftBukkit start
+        Entity entityhuman = this.field_70170_p.func_72856_b(this, 16.0D); // EntityHuman -> Entity
+        EntityTargetEvent event = null;
+
+        if (entityhuman != null && !entityhuman.equals(lastTarget))
+        {
+            event = CraftEventFactory.callEntityTargetEvent(this, entityhuman, EntityTargetEvent.TargetReason.CLOSEST_PLAYER);
+        }
+        else if (lastTarget != null && entityhuman == null)
+        {
+            event = CraftEventFactory.callEntityTargetEvent(this, entityhuman, EntityTargetEvent.TargetReason.FORGOT_TARGET);
+        }
+
+        if (event != null && !event.isCancelled())
+        {
+            entityhuman = event.getTarget() == null ? null : ((CraftEntity) event.getTarget()).getHandle();
+        }
+
+        this.lastTarget = entityhuman;
+        // CraftBukkit end
+
+        if (entityhuman != null)
+        {
+            this.func_70625_a(entityhuman, 10.0F, 20.0F);
         }
 
         if (this.field_70122_E && this.field_70810_d-- <= 0)
         {
             this.field_70810_d = this.func_70806_k();
 
-            if (entityplayer != null)
+            if (entityhuman != null)
             {
                 this.field_70810_d /= 3;
             }
@@ -180,6 +208,21 @@
         if (!this.field_70170_p.field_72995_K && i > 1 && this.func_110143_aJ() <= 0.0F)
         {
             int j = 2 + this.field_70146_Z.nextInt(3);
+            // CraftBukkit start
+            SlimeSplitEvent event = new SlimeSplitEvent((org.bukkit.entity.Slime) this.getBukkitEntity(), j);
+            this.field_70170_p.getServer().getPluginManager().callEvent(event);
+
+            if (!event.isCancelled() && event.getCount() > 0)
+            {
+                j = event.getCount();
+            }
+            else
+            {
+                super.func_70106_y();
+                return;
+            }
+
+            // CraftBukkit end
 
             for (int k = 0; k < j; ++k)
             {
@@ -188,7 +231,7 @@
                 EntitySlime entityslime = this.func_70802_j();
                 entityslime.func_70799_a(i / 2);
                 entityslime.func_70012_b(this.field_70165_t + (double)f, this.field_70163_u + 0.5D, this.field_70161_v + (double)f1, this.field_70146_Z.nextFloat() * 360.0F, 0.0F);
-                this.field_70170_p.func_72838_d(entityslime);
+                this.field_70170_p.addEntity(entityslime, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SLIME_SPLIT); // CraftBukkit - SpawnReason
             }
         }
 
