--- ../net/minecraft/entity/monster/EntityPigZombie.java
+++ ../net/minecraft/entity/monster/EntityPigZombie.java
@@ -14,11 +14,16 @@
 import net.minecraft.util.DamageSource;
 import net.minecraft.world.World;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.event.entity.EntityTargetEvent;
+// CraftBukkit end
+
 public class EntityPigZombie extends EntityZombie
 {
     private static final UUID field_110189_bq = UUID.fromString("49455A49-7EC5-45BA-B886-3B90B23A1718");
     private static final AttributeModifier field_110190_br = (new AttributeModifier(field_110189_bq, "Attacking speed boost", 0.45D, 0)).func_111168_a(false);
-    private int field_70837_d;
+    public int field_70837_d; // CraftBukkit - private -> public
     private int field_70838_e;
     private Entity field_110191_bu;
 
@@ -120,6 +125,24 @@
 
     private void func_70835_c(Entity p_70835_1_)
     {
+        // CraftBukkit start
+        org.bukkit.entity.Entity bukkitTarget = p_70835_1_ == null ? null : p_70835_1_.getBukkitEntity();
+        EntityTargetEvent event = new EntityTargetEvent(this.getBukkitEntity(), bukkitTarget, EntityTargetEvent.TargetReason.PIG_ZOMBIE_TARGET);
+        this.field_70170_p.getServer().getPluginManager().callEvent(event);
+
+        if (event.isCancelled())
+        {
+            return;
+        }
+
+        if (event.getTarget() == null)
+        {
+            this.field_70789_a = null;
+            return;
+        }
+
+        p_70835_1_ = ((org.bukkit.craftbukkit.entity.CraftEntity) event.getTarget()).getHandle();
+        // CraftBukkit end
         this.field_70789_a = p_70835_1_;
         this.field_70837_d = 400 + this.field_70146_Z.nextInt(400);
         this.field_70838_e = this.field_70146_Z.nextInt(40);
@@ -142,20 +165,40 @@
 
     protected void func_70628_a(boolean p_70628_1_, int p_70628_2_)
     {
+        // CraftBukkit start
+        List<org.bukkit.inventory.ItemStack> loot = new java.util.ArrayList<org.bukkit.inventory.ItemStack>();
         int j = this.field_70146_Z.nextInt(2 + p_70628_2_);
-        int k;
-
-        for (k = 0; k < j; ++k)
-        {
-            this.func_70025_b(Item.field_77737_bm.field_77779_bT, 1);
+
+        if (j > 0)
+        {
+            loot.add(CraftItemStack.asNewCraftStack(Item.field_77737_bm, j));
         }
 
         j = this.field_70146_Z.nextInt(2 + p_70628_2_);
 
-        for (k = 0; k < j; ++k)
-        {
-            this.func_70025_b(Item.field_77733_bq.field_77779_bT, 1);
-        }
+        if (j > 0)
+        {
+            loot.add(CraftItemStack.asNewCraftStack(Item.field_77733_bq, j));
+        }
+
+        // Determine rare item drops and add them to the loot
+        if (this.field_70718_bc > 0)
+        {
+            int k = this.field_70146_Z.nextInt(200) - p_70628_2_;
+
+            if (k < 5)
+            {
+                ItemStack itemstack = this.func_70600_l(k <= 0 ? 1 : 0);
+
+                if (itemstack != null)
+                {
+                    loot.add(CraftItemStack.asCraftMirror(itemstack));
+                }
+            }
+        }
+
+        org.bukkit.craftbukkit.event.CraftEventFactory.callEntityDeathEvent(this, loot);
+        // CraftBukkit end
     }
 
     public boolean func_70085_c(EntityPlayer p_70085_1_)
@@ -163,10 +206,12 @@
         return false;
     }
 
-    protected void func_70600_l(int p_70600_1_)
-    {
-        this.func_70025_b(Item.field_77717_p.field_77779_bT, 1);
-    }
+    // CraftBukkit start - Return rare dropped item instead of dropping it
+    protected ItemStack func_70600_l(int i)
+    {
+        return new ItemStack(Item.field_77717_p.field_77779_bT, 1, 0);
+    }
+    // CraftBukkit end
 
     protected int func_70633_aT()
     {
