--- ../net/minecraft/entity/monster/EntityZombie.java
+++ ../net/minecraft/entity/monster/EntityZombie.java
@@ -29,9 +29,16 @@
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.potion.Potion;
 import net.minecraft.potion.PotionEffect;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.util.DamageSource;
 import net.minecraft.util.MathHelper;
 import net.minecraft.world.World;
+
+//CraftBukkit start
+import org.bukkit.craftbukkit.entity.CraftLivingEntity;
+import org.bukkit.event.entity.EntityCombustByEntityEvent;
+import org.bukkit.event.entity.EntityCombustEvent;
+//CraftBukkit end
 
 public class EntityZombie extends EntityMob
 {
@@ -39,6 +46,7 @@
     private static final UUID field_110187_bq = UUID.fromString("B9766B59-9566-4402-BC1F-2EE2A276D836");
     private static final AttributeModifier field_110188_br = new AttributeModifier(field_110187_bq, "Baby speed boost", 0.5D, 0);
     private int field_82234_d;
+    private int lastTick = MinecraftServer.currentTick; // CraftBukkit
 
     public EntityZombie(World p_i1745_1_)
     {
@@ -152,7 +160,16 @@
 
                 if (flag)
                 {
-                    this.func_70015_d(8);
+                    // CraftBukkit start
+                    EntityCombustEvent event = new EntityCombustEvent(this.getBukkitEntity(), 8);
+                    this.field_70170_p.getServer().getPluginManager().callEvent(event);
+
+                    if (!event.isCancelled())
+                    {
+                        this.func_70015_d(event.getDuration());
+                    }
+
+                    // CraftBukkit end
                 }
             }
         }
@@ -219,6 +236,11 @@
         if (!this.field_70170_p.field_72995_K && this.func_82230_o())
         {
             int i = this.func_82233_q();
+            // CraftBukkit start - Use wall time instead of ticks for villager conversion
+            int elapsedTicks = MinecraftServer.currentTick - this.lastTick;
+            this.lastTick = MinecraftServer.currentTick;
+            i *= elapsedTicks;
+            // CraftBukkit end
             this.field_82234_d -= i;
 
             if (this.field_82234_d <= 0)
@@ -236,7 +258,16 @@
 
         if (flag && this.func_70694_bm() == null && this.func_70027_ad() && this.field_70146_Z.nextFloat() < (float)this.field_70170_p.field_73013_u * 0.3F)
         {
-            p_70652_1_.func_70015_d(2 * this.field_70170_p.field_73013_u);
+            // CraftBukkit start
+            EntityCombustByEntityEvent event = new EntityCombustByEntityEvent(this.getBukkitEntity(), p_70652_1_.getBukkitEntity(), 2 * this.field_70170_p.field_73013_u);
+            this.field_70170_p.getServer().getPluginManager().callEvent(event);
+
+            if (!event.isCancelled())
+            {
+                p_70652_1_.func_70015_d(event.getDuration());
+            }
+
+            // CraftBukkit end
         }
 
         return flag;
@@ -272,20 +303,22 @@
         return EnumCreatureAttribute.UNDEAD;
     }
 
-    protected void func_70600_l(int p_70600_1_)
+    // CraftBukkit start - Return rare dropped item instead of dropping it
+    protected ItemStack func_70600_l(int i)
     {
         switch (this.field_70146_Z.nextInt(3))
         {
             case 0:
-                this.func_70025_b(Item.field_77703_o.field_77779_bT, 1);
-                break;
+                return new ItemStack(Item.field_77703_o.field_77779_bT, 1, 0);
             case 1:
-                this.func_70025_b(Item.field_82797_bK.field_77779_bT, 1);
-                break;
+                return new ItemStack(Item.field_82797_bK.field_77779_bT, 1, 0);
             case 2:
-                this.func_70025_b(Item.field_82794_bL.field_77779_bT, 1);
-        }
-    }
+                return new ItemStack(Item.field_82794_bL.field_77779_bT, 1, 0);
+            default:
+                return null;
+        }
+    }
+    // CraftBukkit end
 
     protected void func_82164_bB()
     {
