--- ../net/minecraft/entity/monster/EntitySpider.java
+++ ../net/minecraft/entity/monster/EntitySpider.java
@@ -1,5 +1,6 @@
 package net.minecraft.entity.monster;
 
+import org.bukkit.event.entity.EntityTargetEvent; // CraftBukkit
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityLivingData;
 import net.minecraft.entity.EnumCreatureAttribute;
@@ -83,7 +84,25 @@
 
         if (f1 > 0.5F && this.field_70146_Z.nextInt(100) == 0)
         {
-            this.field_70789_a = null;
+            // CraftBukkit start
+            EntityTargetEvent event = new EntityTargetEvent(this.getBukkitEntity(), null, EntityTargetEvent.TargetReason.FORGOT_TARGET);
+            this.field_70170_p.getServer().getPluginManager().callEvent(event);
+
+            if (!event.isCancelled())
+            {
+                if (event.getTarget() == null)
+                {
+                    this.field_70789_a = null;
+                }
+                else
+                {
+                    this.field_70789_a = ((org.bukkit.craftbukkit.entity.CraftEntity) event.getTarget()).getHandle();
+                }
+
+                return;
+            }
+
+            // CraftBukkit end
         }
         else
         {
@@ -113,12 +132,27 @@
 
     protected void func_70628_a(boolean p_70628_1_, int p_70628_2_)
     {
-        super.func_70628_a(p_70628_1_, p_70628_2_);
+        // CraftBukkit start - Whole method; adapted from super.dropDeathLoot.
+        java.util.List<org.bukkit.inventory.ItemStack> loot = new java.util.ArrayList<org.bukkit.inventory.ItemStack>();
+        int k = this.field_70146_Z.nextInt(3);
+
+        if (p_70628_2_ > 0)
+        {
+            k += this.field_70146_Z.nextInt(p_70628_2_ + 1);
+        }
+
+        if (k > 0)
+        {
+            loot.add(new org.bukkit.inventory.ItemStack(Item.field_77683_K.field_77779_bT, k));
+        }
 
         if (p_70628_1_ && (this.field_70146_Z.nextInt(3) == 0 || this.field_70146_Z.nextInt(1 + p_70628_2_) > 0))
         {
-            this.func_70025_b(Item.field_77728_bu.field_77779_bT, 1);
-        }
+            loot.add(new org.bukkit.inventory.ItemStack(Item.field_77728_bu.field_77779_bT, 1));
+        }
+
+        org.bukkit.craftbukkit.event.CraftEventFactory.callEntityDeathEvent(this, loot); // raise event even for those times when the entity does not drop loot
+        // CraftBukkit end
     }
 
     public boolean func_70617_f_()
@@ -161,7 +195,7 @@
 
     public EntityLivingData func_110161_a(EntityLivingData p_110161_1_)
     {
-        Object p_110161_1_1 = super.func_110161_a(p_110161_1_);
+        Object object = super.func_110161_a(p_110161_1_);
 
         if (this.field_70170_p.field_73012_v.nextInt(100) == 0)
         {
@@ -172,19 +206,19 @@
             entityskeleton.func_70078_a(this);
         }
 
-        if (p_110161_1_1 == null)
-        {
-            p_110161_1_1 = new SpiderEffectsGroupData();
+        if (object == null)
+        {
+            object = new SpiderEffectsGroupData();
 
             if (this.field_70170_p.field_73013_u > 2 && this.field_70170_p.field_73012_v.nextFloat() < 0.1F * this.field_70170_p.func_110746_b(this.field_70165_t, this.field_70163_u, this.field_70161_v))
             {
-                ((SpiderEffectsGroupData)p_110161_1_1).func_111104_a(this.field_70170_p.field_73012_v);
-            }
-        }
-
-        if (p_110161_1_1 instanceof SpiderEffectsGroupData)
-        {
-            int i = ((SpiderEffectsGroupData)p_110161_1_1).field_111105_a;
+                ((SpiderEffectsGroupData) object).func_111104_a(this.field_70170_p.field_73012_v);
+            }
+        }
+
+        if (object instanceof SpiderEffectsGroupData)
+        {
+            int i = ((SpiderEffectsGroupData) object).field_111105_a;
 
             if (i > 0 && Potion.field_76425_a[i] != null)
             {
@@ -192,6 +226,6 @@
             }
         }
 
-        return (EntityLivingData)p_110161_1_1;
+        return (EntityLivingData) object;
     }
 }
