--- ../net/minecraft/block/BlockCocoa.java
+++ ../net/minecraft/block/BlockCocoa.java
@@ -13,6 +13,8 @@
 
 public class BlockCocoa extends BlockDirectional
 {
+    public static final String[] a = new String[] { "cocoa_0", "cocoa_1", "cocoa_2"};
+
     public BlockCocoa(int p_i2181_1_)
     {
         super(p_i2181_1_, Material.field_76254_j);
@@ -34,7 +36,7 @@
             if (i1 < 2)
             {
                 ++i1;
-                p_71847_1_.func_72921_c(p_71847_2_, p_71847_3_, p_71847_4_, i1 << 2 | func_72217_d(l), 2);
+                org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(p_71847_1_, p_71847_2_, p_71847_3_, p_71847_4_, this.field_71990_ca, i1 << 2 | func_72217_d(l)); // CraftBukkit
             }
         }
     }
