--- ../net/minecraft/block/BlockDropper.java
+++ ../net/minecraft/block/BlockDropper.java
@@ -1,5 +1,9 @@
 package net.minecraft.block;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.event.inventory.InventoryMoveItemEvent;
+import org.bukkit.inventory.Inventory;
 import net.minecraft.dispenser.BehaviorDefaultDispenseItem;
 import net.minecraft.dispenser.IBehaviorDispenseItem;
 import net.minecraft.inventory.IInventory;
@@ -10,6 +14,7 @@
 import net.minecraft.tileentity.TileEntityHopper;
 import net.minecraft.util.Facing;
 import net.minecraft.world.World;
+// CraftBukkit end
 
 public class BlockDropper extends BlockDispenser
 {
@@ -30,7 +35,7 @@
         return new TileEntityDropper();
     }
 
-    protected void func_82526_n(World p_82526_1_, int p_82526_2_, int p_82526_3_, int p_82526_4_)
+    public void func_82526_n(World p_82526_1_, int p_82526_2_, int p_82526_3_, int p_82526_4_)   // CraftBukkit - protected -> public
     {
         BlockSourceImpl blocksourceimpl = new BlockSourceImpl(p_82526_1_, p_82526_2_, p_82526_3_, p_82526_4_);
         TileEntityDispenser tileentitydispenser = (TileEntityDispenser)blocksourceimpl.func_82619_j();
@@ -52,10 +57,22 @@
 
                 if (iinventory != null)
                 {
-                    itemstack1 = TileEntityHopper.func_94117_a(iinventory, itemstack.func_77946_l().func_77979_a(1), Facing.field_71588_a[i1]);
+                    // CraftBukkit start - fire event when pushing items into other inventories
+                    CraftItemStack oitemstack = CraftItemStack.asCraftMirror(itemstack.func_77946_l().func_77979_a(1));
+                    Inventory destinationInventory = iinventory.getOwner() != null ? iinventory.getOwner().getInventory() : null;
+                    InventoryMoveItemEvent event = new InventoryMoveItemEvent(tileentitydispenser.getOwner().getInventory(), oitemstack.clone(), destinationInventory, true);
+                    p_82526_1_.getServer().getPluginManager().callEvent(event);
 
-                    if (itemstack1 == null)
+                    if (event.isCancelled())
                     {
+                        return;
+                    }
+
+                    itemstack1 = TileEntityHopper.func_94117_a(iinventory, CraftItemStack.asNMSCopy(event.getItem()), Facing.field_71588_a[i1]);
+
+                    if (event.getItem().equals(oitemstack) && itemstack1 == null)
+                    {
+                        // CraftBukkit end
                         itemstack1 = itemstack.func_77946_l();
 
                         if (--itemstack1.field_77994_a == 0)
