--- ../net/minecraft/block/BlockStationary.java
+++ ../net/minecraft/block/BlockStationary.java
@@ -4,6 +4,8 @@
 import net.minecraft.block.material.Material;
 import net.minecraft.world.IBlockAccess;
 import net.minecraft.world.World;
+
+import org.bukkit.craftbukkit.event.CraftEventFactory;  // CraftBukkit
 
 public class BlockStationary extends BlockFluid
 {
@@ -47,6 +49,11 @@
             int l = p_71847_5_.nextInt(3);
             int i1;
             int j1;
+            // CraftBukkit start - prevent lava putting something on fire, remember igniter block coords
+            int x = p_71847_2_;
+            int y = p_71847_3_;
+            int z = p_71847_4_;
+            // CraftBukkit end
 
             for (i1 = 0; i1 < l; ++i1)
             {
@@ -59,6 +66,16 @@
                 {
                     if (this.func_72216_n(p_71847_1_, p_71847_2_ - 1, p_71847_3_, p_71847_4_) || this.func_72216_n(p_71847_1_, p_71847_2_ + 1, p_71847_3_, p_71847_4_) || this.func_72216_n(p_71847_1_, p_71847_2_, p_71847_3_, p_71847_4_ - 1) || this.func_72216_n(p_71847_1_, p_71847_2_, p_71847_3_, p_71847_4_ + 1) || this.func_72216_n(p_71847_1_, p_71847_2_, p_71847_3_ - 1, p_71847_4_) || this.func_72216_n(p_71847_1_, p_71847_2_, p_71847_3_ + 1, p_71847_4_))
                     {
+                        // CraftBukkit start - prevent lava putting something on fire
+                        if (p_71847_1_.func_72798_a(p_71847_2_, p_71847_3_, p_71847_4_) != Block.field_72067_ar.field_71990_ca)
+                        {
+                            if (CraftEventFactory.callBlockIgniteEvent(p_71847_1_, p_71847_2_, p_71847_3_, p_71847_4_, x, y, z).isCancelled())
+                            {
+                                continue;
+                            }
+                        }
+
+                        // CraftBukkit end
                         p_71847_1_.func_94575_c(p_71847_2_, p_71847_3_, p_71847_4_, Block.field_72067_ar.field_71990_ca);
                         return;
                     }
@@ -81,6 +98,16 @@
 
                     if (p_71847_1_.func_72799_c(p_71847_2_, p_71847_3_ + 1, p_71847_4_) && this.func_72216_n(p_71847_1_, p_71847_2_, p_71847_3_, p_71847_4_))
                     {
+                        // CraftBukkit start - prevent lava putting something on fire
+                        if (p_71847_1_.func_72798_a(p_71847_2_, p_71847_3_ + 1, p_71847_4_) != Block.field_72067_ar.field_71990_ca)
+                        {
+                            if (CraftEventFactory.callBlockIgniteEvent(p_71847_1_, p_71847_2_, p_71847_3_ + 1, p_71847_4_, x, y, z).isCancelled())
+                            {
+                                continue;
+                            }
+                        }
+
+                        // CraftBukkit end
                         p_71847_1_.func_94575_c(p_71847_2_, p_71847_3_ + 1, p_71847_4_, Block.field_72067_ar.field_71990_ca);
                     }
                 }
