--- ../net/minecraft/block/BlockCrops.java
+++ ../net/minecraft/block/BlockCrops.java
@@ -40,8 +40,7 @@
 
                 if (p_71847_5_.nextInt((int)(25.0F / f) + 1) == 0)
                 {
-                    ++l;
-                    p_71847_1_.func_72921_c(p_71847_2_, p_71847_3_, p_71847_4_, l, 2);
+                    org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(p_71847_1_, p_71847_2_, p_71847_3_, p_71847_4_, this.field_71990_ca, ++l); // CraftBukkit
                 }
             }
         }
