--- ../net/minecraft/potion/PotionEffect.java
+++ ../net/minecraft/potion/PotionEffect.java
@@ -2,7 +2,7 @@
 
 import net.minecraft.entity.EntityLiving;
 import net.minecraft.nbt.NBTTagCompound;
-
+import net.minecraft.server.MinecraftServer;
 public class PotionEffect
 {
     private int field_76462_a;
@@ -10,6 +10,7 @@
     private int field_76461_c;
     private boolean field_82723_d;
     private boolean field_82724_e;
+    private int lastTick = MinecraftServer.currentTick; // CraftBukkit
 
     public PotionEffect(int p_i5051_1_, int p_i5051_2_)
     {
@@ -105,7 +106,12 @@
 
     private int func_76454_e()
     {
-        return --this.field_76460_b;
+        // CraftBukkit start - Use wall time instead of ticks for potion effects
+        int elapsedTicks = Math.max(1, MinecraftServer.currentTick - this.lastTick);
+        this.lastTick = MinecraftServer.currentTick;
+        this.field_76460_b -= elapsedTicks;
+        return this.field_76460_b;
+        // CraftBukkit end
     }
 
     public void func_76457_b(EntityLiving p_76457_1_)
