--- ../net/minecraft/network/NetLoginHandler.java
+++ ../net/minecraft/network/NetLoginHandler.java
@@ -1,6 +1,5 @@
 package net.minecraft.network;
 
-import java.io.IOException;
 import java.io.Serializable;
 import java.net.InetAddress;
 import java.net.Socket;
@@ -34,18 +33,26 @@
     public final TcpConnection field_72538_b;
     public boolean field_72539_c;
     private int field_72535_g;
-    public String field_72543_h;
+    private String field_72543_h;
     private volatile boolean field_72544_i;
-    private String field_72541_j = "";
+    private String field_72541_j = Long.toString(field_72537_e.nextLong(), 16); // CraftBukkit - Security fix
     private boolean field_92079_k;
     private SecretKey field_72542_k;
-
-    public NetLoginHandler(MinecraftServer p_i1527_1_, Socket p_i1527_2_, String p_i1527_3_) throws IOException
+    public String hostname = ""; // CraftBukkit - add field
+
+    public NetLoginHandler(MinecraftServer p_i1527_1_, Socket p_i1527_2_, String p_i1527_3_) throws java.io.IOException   // CraftBukkit - throws IOException
     {
         this.field_72534_f = p_i1527_1_;
         this.field_72538_b = new TcpConnection(p_i1527_1_.func_98033_al(), p_i1527_2_, p_i1527_3_, this, p_i1527_1_.func_71250_E().getPrivate());
         this.field_72538_b.field_74468_e = 0;
     }
+
+    // CraftBukkit start
+    public Socket getSocket()
+    {
+        return this.field_72538_b.func_74452_g();
+    }
+    // CraftBukkit end
 
     public void func_72532_c()
     {
@@ -146,7 +153,7 @@
 
             if (this.field_72534_f.func_71266_T())
             {
-                (new ThreadLoginVerifier(this)).start();
+                (new ThreadLoginVerifier(this, field_72534_f.server)).start(); // CraftBukkit - add CraftServer
             }
             else
             {
@@ -159,19 +166,22 @@
 
     public void func_72529_d()
     {
-        String s = this.field_72534_f.func_71203_ab().func_72399_a(this.field_72538_b.func_74430_c(), this.field_72543_h);
-
-        if (s != null)
-        {
-            this.func_72527_a(s);
+        // CraftBukkit start
+        EntityPlayerMP s = this.field_72534_f.func_71203_ab().attemptLogin(this, this.field_72543_h, this.hostname);
+
+        if (s == null)
+        {
+            // this.disconnect(s);
+            return;
+            // CraftBukkit end
         }
         else
         {
-            EntityPlayerMP entityplayermp = this.field_72534_f.func_71203_ab().func_72366_a(this.field_72543_h);
+            EntityPlayerMP entityplayermp = this.field_72534_f.func_71203_ab().processLogin(s); // CraftBukkit - this.g -> s
 
             if (entityplayermp != null)
             {
-                this.field_72534_f.func_71203_ab().func_72355_a(this.field_72538_b, entityplayermp);
+                this.field_72534_f.func_71203_ab().func_72355_a((INetworkManager) this.field_72538_b, entityplayermp);
             }
         }
 
@@ -186,33 +196,45 @@
 
     public void func_72467_a(Packet254ServerPing p_72467_1_)
     {
+        if (this.field_72538_b.func_74452_g() == null)
+        {
+            return;    // CraftBukkit - fix NPE when a client queries a server that is unable to handle it.
+        }
+
         try
         {
             ServerConfigurationManager serverconfigurationmanager = this.field_72534_f.func_71203_ab();
             String s = null;
+            // CraftBukkit
+            org.bukkit.event.server.ServerListPingEvent pingEvent = org.bukkit.craftbukkit.event.CraftEventFactory.callServerListPingEvent(this.field_72534_f.server, getSocket().getInetAddress(), this.field_72534_f.func_71273_Y(), serverconfigurationmanager.func_72394_k(), serverconfigurationmanager.func_72352_l());
 
             if (p_72467_1_.func_140050_d())
             {
-                s = this.field_72534_f.func_71273_Y() + "\u00a7" + serverconfigurationmanager.func_72394_k() + "\u00a7" + serverconfigurationmanager.func_72352_l();
+                // CraftBukkit
+                s = pingEvent.getMotd() + "\u00A7" + serverconfigurationmanager.func_72394_k() + "\u00A7" + pingEvent.getMaxPlayers();
             }
             else
             {
-                List list = Arrays.asList(new Serializable[] {Integer.valueOf(1), Integer.valueOf(78), this.field_72534_f.func_71249_w(), this.field_72534_f.func_71273_Y(), Integer.valueOf(serverconfigurationmanager.func_72394_k()), Integer.valueOf(serverconfigurationmanager.func_72352_l())});
-                Object object;
-
-                for (Iterator iterator = list.iterator(); iterator.hasNext(); s = s + object.toString().replaceAll("\u0000", ""))
+                // CraftBukkit start - Don't create a list from an array
+                Object[] list = new Object[] { 1, 78, this.field_72534_f.func_71249_w(), pingEvent.getMotd(), serverconfigurationmanager.func_72394_k(), pingEvent.getMaxPlayers() };
+                StringBuilder builder = new StringBuilder();
+
+                for (Object object : list)
                 {
-                    object = iterator.next();
-
-                    if (s == null)
+                    if (builder.length() == 0)
                     {
-                        s = "\u00a7";
+                        builder.append('\u00A7');
                     }
                     else
                     {
-                        s = s + "\u0000";
+                        builder.append('\0');
                     }
+
+                    builder.append(org.apache.commons.lang.StringUtils.replace(object.toString(), "\0", ""));
                 }
+
+                s = builder.toString();
+                // CraftBukkit end
             }
 
             InetAddress inetaddress = null;
@@ -278,7 +300,7 @@
         return p_72533_0_.field_72543_h;
     }
 
-    public static boolean func_72531_a(NetLoginHandler p_72531_0_, boolean p_72531_1_)
+    static boolean func_72531_a(NetLoginHandler p_72531_0_, boolean p_72531_1_)
     {
         return p_72531_0_.field_72544_i = p_72531_1_;
     }
