--- ../net/minecraft/item/ItemDye.java
+++ ../net/minecraft/item/ItemDye.java
@@ -1,5 +1,8 @@
 package net.minecraft.item;
 
+// CraftBukkit start
+import org.bukkit.entity.Player;
+import org.bukkit.event.entity.SheepDyeWoolEvent;
 import net.minecraft.block.Block;
 import net.minecraft.block.BlockCocoa;
 import net.minecraft.block.BlockColored;
@@ -13,8 +16,10 @@
 import net.minecraft.entity.EntityLivingBase;
 import net.minecraft.entity.passive.EntitySheep;
 import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.entity.player.EntityPlayerMP;
 import net.minecraft.util.MathHelper;
 import net.minecraft.world.World;
+// CraftBukkit end
 
 public class ItemDye extends Item
 {
@@ -46,7 +51,7 @@
         {
             if (p_77648_1_.func_77960_j() == 15)
             {
-                if (func_96604_a(p_77648_1_, p_77648_3_, p_77648_4_, p_77648_5_, p_77648_6_))
+                if (func_96604_a(p_77648_1_, p_77648_3_, p_77648_4_, p_77648_5_, p_77648_6_, p_77648_2_))   // CraftBukkit - pass entity for StructureGrowEvent
                 {
                     if (!p_77648_3_.field_72995_K)
                     {
@@ -112,8 +117,15 @@
         }
     }
 
+    // CraftBukkit start
     public static boolean func_96604_a(ItemStack p_96604_0_, World p_96604_1_, int p_96604_2_, int p_96604_3_, int p_96604_4_)
     {
+        return func_96604_a(p_96604_0_, p_96604_1_, p_96604_2_, p_96604_3_, p_96604_4_, null);
+    }
+
+    public static boolean func_96604_a(ItemStack p_96604_0_, World p_96604_1_, int p_96604_2_, int p_96604_3_, int p_96604_4_, EntityPlayer player)
+    {
+        // CraftBukkit end
         int l = p_96604_1_.func_72798_a(p_96604_2_, p_96604_3_, p_96604_4_);
 
         if (l == Block.field_71987_y.field_71990_ca)
@@ -122,7 +134,10 @@
             {
                 if ((double)p_96604_1_.field_73012_v.nextFloat() < 0.45D)
                 {
-                    ((BlockSapling)Block.field_71987_y).func_96477_c(p_96604_1_, p_96604_2_, p_96604_3_, p_96604_4_, p_96604_1_.field_73012_v);
+                    // CraftBukkit start
+                    Player player = (player instanceof EntityPlayerMP) ? (Player) player.getBukkitEntity() : null;
+                    ((BlockSapling) Block.field_71987_y).func_72269_c(p_96604_1_, p_96604_2_, p_96604_3_, p_96604_4_, p_96604_1_.field_73012_v, true, player, null);
+                    // CraftBukkit end
                 }
 
                 --p_96604_0_.field_77994_a;
@@ -257,10 +272,13 @@
             {
                 if ((double)p_96604_1_.field_73012_v.nextFloat() < 0.4D)
                 {
-                    ((BlockMushroom)Block.field_71973_m[l]).func_72271_c(p_96604_1_, p_96604_2_, p_96604_3_, p_96604_4_, p_96604_1_.field_73012_v);
-                }
-
-                --p_96604_0_.field_77994_a;
+                    // CraftBukkit start - Validate
+                    Player player = (player instanceof EntityPlayerMP) ? (Player) player.getBukkitEntity() : null;
+                    ((BlockMushroom) Block.field_71973_m[l]).func_72271_c(p_96604_1_, p_96604_2_, p_96604_3_, p_96604_4_, p_96604_1_.field_73012_v, true, player, p_96604_0_);
+                }
+
+                //--itemstack.count; - called later if the bonemeal attempt was not cancelled by a plugin
+                // CraftBukkit end
             }
 
             return true;
@@ -276,6 +294,18 @@
 
             if (!entitysheep.func_70892_o() && entitysheep.func_70896_n() != i)
             {
+                // CraftBukkit start
+                byte bColor = (byte) i;
+                SheepDyeWoolEvent event = new SheepDyeWoolEvent((org.bukkit.entity.Sheep) entitysheep.getBukkitEntity(), org.bukkit.DyeColor.getByData(bColor));
+                entitysheep.field_70170_p.getServer().getPluginManager().callEvent(event);
+
+                if (event.isCancelled())
+                {
+                    return false;
+                }
+
+                i = (byte) event.getColor().getWoolData();
+                // CraftBukkit end
                 entitysheep.func_70891_b(i);
                 --p_111207_1_.field_77994_a;
             }
