--- ../net/minecraft/item/ItemMonsterPlacer.java
+++ ../net/minecraft/item/ItemMonsterPlacer.java
@@ -15,7 +15,6 @@
 import net.minecraft.util.MovingObjectPosition;
 import net.minecraft.util.StatCollector;
 import net.minecraft.world.World;
-
 public class ItemMonsterPlacer extends Item
 {
     public ItemMonsterPlacer(int p_i1906_1_)
@@ -40,7 +39,7 @@
 
     public boolean func_77648_a(ItemStack p_77648_1_, EntityPlayer p_77648_2_, World p_77648_3_, int p_77648_4_, int p_77648_5_, int p_77648_6_, int p_77648_7_, float p_77648_8_, float p_77648_9_, float p_77648_10_)
     {
-        if (p_77648_3_.field_72995_K)
+        if (p_77648_3_.field_72995_K || p_77648_1_.func_77960_j() == 48 || p_77648_1_.func_77960_j() == 49 || p_77648_1_.func_77960_j() == 63 || p_77648_1_.func_77960_j() == 64)   // CraftBukkit
         {
             return true;
         }
@@ -153,7 +152,7 @@
                     entityliving.field_70759_as = entityliving.field_70177_z;
                     entityliving.field_70761_aq = entityliving.field_70177_z;
                     entityliving.func_110161_a((EntityLivingData)null);
-                    p_77840_0_.func_72838_d(entity);
+                    p_77840_0_.addEntity(entity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SPAWNER_EGG); // CraftBukkit
                     entityliving.func_70642_aH();
                 }
             }
