--- ../net/minecraft/item/ItemMap.java
+++ ../net/minecraft/item/ItemMap.java
@@ -1,5 +1,8 @@
 package net.minecraft.item;
 
+// CraftBukkit start
+import org.bukkit.Bukkit;
+import org.bukkit.event.server.MapInitializeEvent;
 import net.minecraft.block.Block;
 import net.minecraft.block.material.MapColor;
 import net.minecraft.entity.Entity;
@@ -8,9 +11,12 @@
 import net.minecraft.network.packet.Packet131MapData;
 import net.minecraft.util.MathHelper;
 import net.minecraft.world.World;
+import net.minecraft.world.WorldSavedData;
+import net.minecraft.world.WorldServer;
 import net.minecraft.world.chunk.Chunk;
 import net.minecraft.world.storage.MapData;
 import net.minecraft.world.storage.MapInfo;
+// CraftBukkit end
 
 public class ItemMap extends ItemMapBase
 {
@@ -22,21 +28,26 @@
 
     public MapData func_77873_a(ItemStack p_77873_1_, World p_77873_2_)
     {
+        World worldMain = p_77873_2_.getServer().getServer().worlds.get(0); // CraftBukkit - store reference to primary world
         String s = "map_" + p_77873_1_.func_77960_j();
-        MapData mapdata = (MapData)p_77873_2_.func_72943_a(MapData.class, s);
+        MapData mapdata = (MapData) worldMain.func_72943_a(MapData.class, s); // CraftBukkit - use primary world for maps
 
         if (mapdata == null && !p_77873_2_.field_72995_K)
         {
-            p_77873_1_.func_77964_b(p_77873_2_.func_72841_b("map"));
+            p_77873_1_.func_77964_b(worldMain.func_72841_b("map")); // CraftBukkit - use primary world for maps
             s = "map_" + p_77873_1_.func_77960_j();
             mapdata = new MapData(s);
             mapdata.field_76197_d = 3;
             int i = 128 * (1 << mapdata.field_76197_d);
             mapdata.field_76201_a = Math.round((float)p_77873_2_.func_72912_H().func_76079_c() / (float)i) * i;
             mapdata.field_76199_b = Math.round((float)(p_77873_2_.func_72912_H().func_76074_e() / i)) * i;
-            mapdata.field_76200_c = (byte)p_77873_2_.field_73011_w.field_76574_g;
+            mapdata.field_76200_c = (byte)((WorldServer) p_77873_2_).dimension;  // CraftBukkit - fixes Bukkit multiworld maps
             mapdata.func_76185_a();
-            p_77873_2_.func_72823_a(s, mapdata);
+            worldMain.func_72823_a(s, (WorldSavedData) mapdata); // CraftBukkit - use primary world for maps
+            // CraftBukkit start
+            MapInitializeEvent event = new MapInitializeEvent(mapdata.mapView);
+            Bukkit.getServer().getPluginManager().callEvent(event);
+            // CraftBukkit end
         }
 
         return mapdata;
@@ -44,7 +55,8 @@
 
     public void func_77872_a(World p_77872_1_, Entity p_77872_2_, MapData p_77872_3_)
     {
-        if (p_77872_1_.field_73011_w.field_76574_g == p_77872_3_.field_76200_c && p_77872_2_ instanceof EntityPlayer)
+        // CraftBukkit
+        if (((WorldServer) p_77872_1_).dimension == p_77872_3_.field_76200_c && p_77872_2_ instanceof EntityPlayer)
         {
             short short1 = 128;
             short short2 = 128;
@@ -288,6 +300,7 @@
         if (p_77622_1_.func_77942_o() && p_77622_1_.func_77978_p().func_74767_n("map_is_scaling"))
         {
             MapData mapdata = Item.field_77744_bd.func_77873_a(p_77622_1_, p_77622_2_);
+            p_77622_2_ = p_77622_2_.getServer().getServer().worlds.get(0); // CraftBukkit - use primary world for maps
             p_77622_1_.func_77964_b(p_77622_2_.func_72841_b("map"));
             MapData mapdata1 = new MapData("map_" + p_77622_1_.func_77960_j());
             mapdata1.field_76197_d = (byte)(mapdata.field_76197_d + 1);
@@ -301,7 +314,11 @@
             mapdata1.field_76199_b = mapdata.field_76199_b;
             mapdata1.field_76200_c = mapdata.field_76200_c;
             mapdata1.func_76185_a();
-            p_77622_2_.func_72823_a("map_" + p_77622_1_.func_77960_j(), mapdata1);
+            p_77622_2_.func_72823_a("map_" + p_77622_1_.func_77960_j(), (WorldSavedData) mapdata1);
+            // CraftBukkit start
+            MapInitializeEvent event = new MapInitializeEvent(mapdata1.mapView);
+            Bukkit.getServer().getPluginManager().callEvent(event);
+            // CraftBukkit end
         }
     }
 }
